{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "79069c72-cbac-46bc-80ae-05778607e41e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from joblib import Parallel, delayed\n",
    "import pickle\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from Bio import Entrez\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "243b9296-2e3c-4c54-9e3b-b29c68207627",
   "metadata": {},
   "outputs": [],
   "source": [
    "PWD = Path().cwd()\n",
    "DATA = Path(\"../data\")\n",
    "IPS = DATA / \"rnaseq_res_ips\"\n",
    "HIMORNA = DATA / \"himorna_peaks\"\n",
    "FANTOM = DATA / \"fantom_cat\"\n",
    "PREPROCESSING = Path(\"../preprocessing\")\n",
    "HIMORNA_INDEX = PREPROCESSING / \"himorna_index\"\n",
    "IPS_INDEX = PREPROCESSING / \"ips_deg_index\"\n",
    "MATRICES = PREPROCESSING / \"matrices\"\n",
    "FISHER = PREPROCESSING / \"fisher_exact_test\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d699fa7-eba0-473c-9334-6ee02bbbdffd",
   "metadata": {},
   "source": [
    "## himorna peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "44d059c0-7b6f-4986-8c5e-eae300edc757",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2995/237761342.py:8: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  himorna_peaks = pd.concat(himorna_peaks)\n"
     ]
    }
   ],
   "source": [
    "to_peaks = HIMORNA / \"Himorna_all_peaks_with_correlations_ensembl_id.tsv.gz\"\n",
    "himorna_peaks = pd.read_csv(\n",
    "        to_peaks,\n",
    "        chunksize=100000000,\n",
    "        sep=\"\\t\",\n",
    "        names=[\"hm\", \"lnc\", \"lnc_ens\", \"peak\", \"chrom\", \"start\", \"end\", \"gene\", \"gene_ens\", \"corr\"]\n",
    "    )\n",
    "himorna_peaks = pd.concat(himorna_peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3f79be8f-a55b-4ebf-8c5c-491f5acb42f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92208462, 10)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "himorna_peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b6d8004c-8913-4f43-b9b5-eb6c5378848e",
   "metadata": {},
   "outputs": [],
   "source": [
    "himorna_peaks[\"lnc\"] = himorna_peaks[\"lnc\"].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "c8da42b0-d426-44a9-8876-2cba1d473cee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92208462, 10)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "himorna_peaks = himorna_peaks.drop_duplicates()\n",
    "himorna_peaks = himorna_peaks.dropna(subset=[\"hm\",\"lnc_ens\",\"gene_ens\"])\n",
    "himorna_peaks.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "20957506-b8be-42ff-b64a-6736e0a6819d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hm</th>\n",
       "      <th>lnc</th>\n",
       "      <th>lnc_ens</th>\n",
       "      <th>peak</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_ens</th>\n",
       "      <th>corr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>AL390961.3</td>\n",
       "      <td>ENSG00000279212</td>\n",
       "      <td>peak_391</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6203922</td>\n",
       "      <td>6205912</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "      <td>-0.540072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>SUGT1-DT</td>\n",
       "      <td>ENSG00000273723</td>\n",
       "      <td>peak_391</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6203922</td>\n",
       "      <td>6205912</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "      <td>-0.493172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>AC004148.1</td>\n",
       "      <td>ENSG00000263272</td>\n",
       "      <td>peak_391</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6203922</td>\n",
       "      <td>6205912</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "      <td>-0.406290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>CADM3-AS1</td>\n",
       "      <td>ENSG00000225670</td>\n",
       "      <td>peak_391</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6203922</td>\n",
       "      <td>6205912</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "      <td>0.556495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H3K9me3</td>\n",
       "      <td>PCAT19</td>\n",
       "      <td>ENSG00000267107</td>\n",
       "      <td>peak_391</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6203922</td>\n",
       "      <td>6205912</td>\n",
       "      <td>RPL22</td>\n",
       "      <td>ENSG00000116251</td>\n",
       "      <td>0.473327</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        hm         lnc          lnc_ens      peak chrom    start      end  \\\n",
       "0  H3K9me3  AL390961.3  ENSG00000279212  peak_391  chr1  6203922  6205912   \n",
       "1  H3K9me3    SUGT1-DT  ENSG00000273723  peak_391  chr1  6203922  6205912   \n",
       "2  H3K9me3  AC004148.1  ENSG00000263272  peak_391  chr1  6203922  6205912   \n",
       "3  H3K9me3   CADM3-AS1  ENSG00000225670  peak_391  chr1  6203922  6205912   \n",
       "4  H3K9me3      PCAT19  ENSG00000267107  peak_391  chr1  6203922  6205912   \n",
       "\n",
       "    gene         gene_ens      corr  \n",
       "0  RPL22  ENSG00000116251 -0.540072  \n",
       "1  RPL22  ENSG00000116251 -0.493172  \n",
       "2  RPL22  ENSG00000116251 -0.406290  \n",
       "3  RPL22  ENSG00000116251  0.556495  \n",
       "4  RPL22  ENSG00000116251  0.473327  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "himorna_peaks.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e28f6175-db75-4f7e-adcf-cc14c170b4d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: >"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAGsCAYAAAAPJKchAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAvLklEQVR4nO3df3TU1Z3/8deQhAlBEgsYk0iAaAUhqIuJSrT8shIMlcIpp+rCQnTVs2z9nc1hCeqa1KrYUhfdKtEaiEpTOW7Eck4QzR4N0Aa2AqHFAml1I3EhkYVKglAmA7nfP/xmjsNMfsxkJneSeT7OyaGfO/d+5r5zP5/46mc+M+MwxhgBAABYMsj2BAAAQHQjjAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACr+lUY2bZtm+bOnau0tDQ5HA698847AY0vLi6Ww+Hw+Rk6dGh4JgwAALrVr8LIqVOndPXVV+sXv/hFUOMLCwvV1NTk9TNx4kT98Ic/DPFMAQBAT/WrMJKXl6ef/OQn+sEPfuD38ba2Ni1btkyXXHKJhg4dquuvv141NTWexy+44AKlpKR4fr744gvt379fd999dx9VAAAAzhdrewKhdNddd+mzzz7Tm2++qbS0NG3cuFG33HKL9u3bp8svv9yn/6uvvqpx48Zp6tSpFmYLAACkfnZlpCuffvqpfv3rX+utt97S1KlTddlll6mwsFDf+c53tG7dOp/+LpdLv/rVr7gqAgCAZQPmysiePXtkjNG4ceO82l0ul0aMGOHT/+2339bJkye1ZMmSvpoiAADwY8CEkfb2dsXExGj37t2KiYnxeuyCCy7w6f/qq6/q1ltvVUpKSl9NEQAA+DFgwsjkyZN17tw5HT16tNt7QBoaGvThhx9q06ZNfTQ7AADQmX4VRr766it98sknnu2Ghgbt3btXw4cP17hx47Ro0SItWbJEP//5zzV58mQdO3ZMH3zwga688krNmTPHM27t2rVKTU1VXl6ejTIAAMA3OIwxxvYkeqqmpkYzZ870ac/Pz1d5ebncbrd+8pOf6PXXX9fhw4c1YsQI5eTkqKSkRFdeeaWkr1/OGTNmjJYsWaKnnnqqr0sAAADn6VdhBAAADDwD5q29AACgfyKMAAAAq/rFDazt7e06cuSIhg0bJofDYXs6AACgB4wxOnnypNLS0jRoUOfXP/pFGDly5IjS09NtTwMAAATh888/16hRozp9vF+EkWHDhkn6upjExMSg9+N2u/X+++8rNzdXcXFxoZpeRImGGqXoqDMaapSio85oqFGKjjqjoUYpdHW2trYqPT3d89/xzvSLMNLx0kxiYmKvw0hCQoISExMH7EEUDTVK0VFnNNQoRUed0VCjFB11RkONUujr7O4WC25gBQAAVhFGAACAVYQRAABgVUBhZM2aNbrqqqs8927k5OTo3Xff7bR/TU2NHA6Hz8/Bgwd7PXEAADAwBHQD66hRo7Ry5Up9+9vfliS99tprmjdvnurq6pSZmdnpuPr6eq8bTy+66KIgpwsAAAaagMLI3LlzvbafeuoprVmzRjt37uwyjCQnJ+vCCy8MaoIAAGBgC/qtvefOndNbb72lU6dOKScnp8u+kydP1pkzZzRx4kQ99thjfr9595tcLpdcLpdnu7W1VdLXbzVyu93BTtkztjf7iHTRUKMUHXVGQ41SdNQZDTVK0VFnNNQoha7Ono4P+Ft79+3bp5ycHJ05c0YXXHCBKioqNGfOHL996+vrtW3bNmVlZcnlcumNN95QaWmpampqNG3atE6fo7i4WCUlJT7tFRUVSkhICGS6AADAktOnT2vhwoVqaWnp8nPCAg4jbW1tamxs1IkTJ1RZWalXX31VW7du1cSJE3s0fu7cuXI4HNq0aVOnffxdGUlPT9exY8d6/aFn1dXVmjVr1oD9sJpoqFGKjjqjoUYpOuqMhhql6KgzGmqUQldna2urRo4c2W0YCfhlmsGDB3tuYM3OztZHH32k559/Xi+//HKPxk+ZMkXr16/vso/T6ZTT6fRpj4uLC8nih2o/kSwaapSio85oqFGKjjqjoUYpOuqMhhql3tfZ07G9/pwRY4zXVYzu1NXVKTU1tbdPCwAABoiAroysWLFCeXl5Sk9P18mTJ/Xmm2+qpqZGW7ZskSQVFRXp8OHDev311yVJq1ev1tixY5WZmam2tjatX79elZWVqqysDH0lAACgXwoojHzxxRdavHixmpqalJSUpKuuukpbtmzRrFmzJElNTU1qbGz09G9ra1NhYaEOHz6sIUOGKDMzU1VVVZ3e8AoAAKJPQGGkrKysy8fLy8u9tpctW6Zly5YFPCkAABA9gv6cEQBA/zJ2eVVY9++MMfrpddKk4vfkOtf1V8b31GcrvxeS/SCy8UV5AADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsCiiMrFmzRldddZUSExOVmJionJwcvfvuu12O2bp1q7KyshQfH69LL71UpaWlvZowAAAYWAIKI6NGjdLKlSu1a9cu7dq1SzfddJPmzZunP/3pT377NzQ0aM6cOZo6darq6uq0YsUKPfjgg6qsrAzJ5AEAQP8XG0jnuXPnem0/9dRTWrNmjXbu3KnMzEyf/qWlpRo9erRWr14tSZowYYJ27dqlVatWacGCBcHPGgAADBgBhZFvOnfunN566y2dOnVKOTk5fvvs2LFDubm5Xm2zZ89WWVmZ3G634uLi/I5zuVxyuVye7dbWVkmS2+2W2+0Odsqesb3ZR6SLhhql6KgzGmqUoqPOSKnRGWPCu/9BxuvfULD9OztfpKxluIWqzp6OdxhjAjpq9u3bp5ycHJ05c0YXXHCBKioqNGfOHL99x40bpzvvvFMrVqzwtNXW1urGG2/UkSNHlJqa6ndccXGxSkpKfNorKiqUkJAQyHQBAIAlp0+f1sKFC9XS0qLExMRO+wV8ZWT8+PHau3evTpw4ocrKSuXn52vr1q2aOHGi3/4Oh8NruyP7nN/+TUVFRSooKPBst7a2Kj09Xbm5uV0W0x23263q6mrNmjWr06sy/V001ChFR53RUKMUHXVGSo2Tit8L6/6dg4yezG7X47sGydXe+d/4QHxcPDsk+wmVSFnLcAtVnR2vbHQn4DAyePBgffvb35YkZWdn66OPPtLzzz+vl19+2advSkqKmpubvdqOHj2q2NhYjRgxotPncDqdcjqdPu1xcXEhWfxQ7SeSRUONUnTUGQ01StFRp+0aXedCExC6fZ52R8ieK1KPCdtr2Vd6W2dPx/b6c0aMMV73d3xTTk6Oqqurvdref/99ZWdnR8UiAgCA7gUURlasWKHt27frs88+0759+/Too4+qpqZGixYtkvT1yytLlizx9F+6dKkOHTqkgoICHThwQGvXrlVZWZkKCwtDWwUAAOi3AnqZ5osvvtDixYvV1NSkpKQkXXXVVdqyZYtmzZolSWpqalJjY6Onf0ZGhjZv3qxHHnlEL774otLS0vTCCy/wtl4AAOARUBgpKyvr8vHy8nKftunTp2vPnj0BTQoAAEQPvpsGAABYRRgBAABWEUYAAIBVQX8cPBCoscurQrYvZ4zRT6/7+kOcwvnZCZ+t/F7Y9g0A+BpXRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFUBhZFnnnlG1157rYYNG6bk5GTNnz9f9fX1XY6pqamRw+Hw+Tl48GCvJg4AAAaGgMLI1q1bdd9992nnzp2qrq7W2bNnlZubq1OnTnU7tr6+Xk1NTZ6fyy+/POhJAwCAgSM2kM5btmzx2l63bp2Sk5O1e/duTZs2rcuxycnJuvDCCwOeIAAAGNgCCiPna2lpkSQNHz68276TJ0/WmTNnNHHiRD322GOaOXNmp31dLpdcLpdnu7W1VZLkdrvldruDnm/H2N7sI9JFco3OGBO6fQ0yXv+Gi83fYySvZShFQ52RUmMoz0G/+w/DeWn7d3a+SFnLcAtVnT0d7zDGBHXUGGM0b948ffnll9q+fXun/err67Vt2zZlZWXJ5XLpjTfeUGlpqWpqajq9mlJcXKySkhKf9oqKCiUkJAQzXQAA0MdOnz6thQsXqqWlRYmJiZ32CzqM3HfffaqqqtJvf/tbjRo1KqCxc+fOlcPh0KZNm/w+7u/KSHp6uo4dO9ZlMd1xu92qrq7WrFmzFBcXF/R+Ilkk1zip+L2Q7cs5yOjJ7HY9vmuQXO2OkO33fB8Xzw7bvrsTyWsZStFQZ6TUGMpz0J9wnJc2z0F/ImUtwy1Udba2tmrkyJHdhpGgXqZ54IEHtGnTJm3bti3gICJJU6ZM0fr16zt93Ol0yul0+rTHxcWFZPFDtZ9IFok1us6FPjS42h1h2W+HSPgdRuJahkM01Gm7xnCeK17PE8LzMlKPCdtr2Vd6W2dPxwYURowxeuCBB7Rx40bV1NQoIyMjqMnV1dUpNTU1qLEAAGBgCSiM3HfffaqoqNBvfvMbDRs2TM3NzZKkpKQkDRkyRJJUVFSkw4cP6/XXX5ckrV69WmPHjlVmZqba2tq0fv16VVZWqrKyMsSlAACA/iigMLJmzRpJ0owZM7za161bpzvvvFOS1NTUpMbGRs9jbW1tKiws1OHDhzVkyBBlZmaqqqpKc+bM6d3MAQDAgBDwyzTdKS8v99petmyZli1bFtCkAABA9OC7aQAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVgUURp555hlde+21GjZsmJKTkzV//nzV19d3O27r1q3KyspSfHy8Lr30UpWWlgY9YQAAMLAEFEa2bt2q++67Tzt37lR1dbXOnj2r3NxcnTp1qtMxDQ0NmjNnjqZOnaq6ujqtWLFCDz74oCorK3s9eQAA0P/FBtJ5y5YtXtvr1q1TcnKydu/erWnTpvkdU1paqtGjR2v16tWSpAkTJmjXrl1atWqVFixYENysAQDAgBFQGDlfS0uLJGn48OGd9tmxY4dyc3O92mbPnq2ysjK53W7FxcX5jHG5XHK5XJ7t1tZWSZLb7Zbb7Q56vh1je7OPSBfJNTpjTOj2Nch4/RsuNn+PkbyWoRQNdUZKjaE8B/3uPwznpe3f2fkiZS3DLVR19nS8wxgT1FFjjNG8efP05Zdfavv27Z32GzdunO68806tWLHC01ZbW6sbb7xRR44cUWpqqs+Y4uJilZSU+LRXVFQoISEhmOkCAIA+dvr0aS1cuFAtLS1KTEzstF/QV0buv/9+/fGPf9Rvf/vbbvs6HA6v7Y78c357h6KiIhUUFHi2W1tblZ6ertzc3C6L6Y7b7VZ1dbVmzZrl94rMQBDJNU4qfi9k+3IOMnoyu12P7xokV7v/4ygUPi6eHbZ9dyeS1zKUoqHOSKkxlOegP+E4L22eg/5EylqGW6jq7HhloztBhZEHHnhAmzZt0rZt2zRq1Kgu+6akpKi5udmr7ejRo4qNjdWIESP8jnE6nXI6nT7tcXFxIVn8UO0nkkVija5zoQ8NrnZHWPbbIRJ+h5G4luEQDXXarjGc54rX84TwvIzUY8L2WvaV3tbZ07EBvZvGGKP7779fb7/9tj744ANlZGR0OyYnJ0fV1dVebe+//76ys7OjYiEBAEDXAgoj9913n9avX6+KigoNGzZMzc3Nam5u1t/+9jdPn6KiIi1ZssSzvXTpUh06dEgFBQU6cOCA1q5dq7KyMhUWFoauCgAA0G8FFEbWrFmjlpYWzZgxQ6mpqZ6fDRs2ePo0NTWpsbHRs52RkaHNmzerpqZGf/d3f6cnn3xSL7zwAm/rBQAAkgK8Z6Qnb7wpLy/3aZs+fbr27NkTyFMBAIAowXfTAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsCjiMbNu2TXPnzlVaWpocDofeeeedLvvX1NTI4XD4/Bw8eDDYOQMAgAEkNtABp06d0tVXX6277rpLCxYs6PG4+vp6JSYmerYvuuiiQJ8aAAAMQAGHkby8POXl5QX8RMnJybrwwgsDHgcAAAa2gMNIsCZPnqwzZ85o4sSJeuyxxzRz5sxO+7pcLrlcLs92a2urJMntdsvtdgc9h46xvdlHpIvkGp0xJnT7GmS8/g0Xm7/HSF7LUIqGOiOlxlCeg373H4bz0vbv7HyRspbhFqo6ezreYYwJ+qhxOBzauHGj5s+f32mf+vp6bdu2TVlZWXK5XHrjjTdUWlqqmpoaTZs2ze+Y4uJilZSU+LRXVFQoISEh2OkCAIA+dPr0aS1cuFAtLS1et2qcL+xhxJ+5c+fK4XBo06ZNfh/3d2UkPT1dx44d67KY7rjdblVXV2vWrFmKi4sLej+RLJJrnFT8Xsj25Rxk9GR2ux7fNUiudkfI9nu+j4tnh23f3YnktQylaKgzUmoM5TnoTzjOS5vnoD+RspbhFqo6W1tbNXLkyG7DSJ+9TPNNU6ZM0fr16zt93Ol0yul0+rTHxcWFZPFDtZ9IFok1us6FPjS42h1h2W+HSPgdRuJahkM01Gm7xnCeK17PE8LzMlKPCdtr2Vd6W2dPx1r5nJG6ujqlpqbaeGoAABBhAr4y8tVXX+mTTz7xbDc0NGjv3r0aPny4Ro8eraKiIh0+fFivv/66JGn16tUaO3asMjMz1dbWpvXr16uyslKVlZWhqwIAAPRbAYeRXbt2eb0TpqCgQJKUn5+v8vJyNTU1qbGx0fN4W1ubCgsLdfjwYQ0ZMkSZmZmqqqrSnDlzQjB9AADQ3wUcRmbMmKGu7nktLy/32l62bJmWLVsW8MQAAEB04LtpAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGAVYQQAAFhFGAEAAFYRRgAAgFWEEQAAYBVhBAAAWEUYAQAAVhFGAACAVYQRAABgFWEEAABYRRgBAABWEUYAAIBVhBEAAGBVrO0JAEB/NHZ5VY/7OmOMfnqdNKn4PbnOOcI4K6B/4soIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArAo4jGzbtk1z585VWlqaHA6H3nnnnW7HbN26VVlZWYqPj9ell16q0tLSYOYKAAAGoIDDyKlTp3T11VfrF7/4RY/6NzQ0aM6cOZo6darq6uq0YsUKPfjgg6qsrAx4sgAAYOAJ+Ivy8vLylJeX1+P+paWlGj16tFavXi1JmjBhgnbt2qVVq1ZpwYIFgT49AAAYYML+rb07duxQbm6uV9vs2bNVVlYmt9utuLg4nzEul0sul8uz3draKklyu91yu91Bz6VjbG/2EekiuUZnjAndvgYZr3/DxebvMZLXMpT6a52BHM99dbzaFo46I+246K/Ha6BCVWdPxzuMMUEfNQ6HQxs3btT8+fM77TNu3DjdeeedWrFihaettrZWN954o44cOaLU1FSfMcXFxSopKfFpr6ioUEJCQrDTBQAAfej06dNauHChWlpalJiY2Gm/sF8Zkb4OLd/UkX/Ob+9QVFSkgoICz3Zra6vS09OVm5vbZTHdcbvdqq6u1qxZs/xekRkIIrnGScXvhWxfzkFGT2a36/Fdg+Rq938chcLHxbPDtu/uRPJahlJ/rTOQ47mvjlfbwlGnzXPQn/56vAYqVHV2vLLRnbCHkZSUFDU3N3u1HT16VLGxsRoxYoTfMU6nU06n06c9Li4uJIsfqv1Eskis0XUu9H+EXe2OsOy3QyT8DiNxLcOhv9UZzHEX7uM1UoSyzkg9Jvrb8Rqs3tbZ07Fh/5yRnJwcVVdXe7W9//77ys7OjoqFBAAAXQs4jHz11Vfau3ev9u7dK+nrt+7u3btXjY2Nkr5+iWXJkiWe/kuXLtWhQ4dUUFCgAwcOaO3atSorK1NhYWFoKgAAAP1awC/T7Nq1SzNnzvRsd9zbkZ+fr/LycjU1NXmCiSRlZGRo8+bNeuSRR/Tiiy8qLS1NL7zwAm/rBQAAkoIIIzNmzFBXb8ApLy/3aZs+fbr27NkT6FMBAIAowHfTAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKvC/q29QH82dnmVted2xhj99Lqvv6o+kG9A/Wzl98I4KwAIPa6MAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMCqoMLISy+9pIyMDMXHxysrK0vbt2/vtG9NTY0cDofPz8GDB4OeNAAAGDgCDiMbNmzQww8/rEcffVR1dXWaOnWq8vLy1NjY2OW4+vp6NTU1eX4uv/zyoCcNAAAGjoDDyHPPPae7775b99xzjyZMmKDVq1crPT1da9as6XJccnKyUlJSPD8xMTFBTxoAAAwcsYF0bmtr0+7du7V8+XKv9tzcXNXW1nY5dvLkyTpz5owmTpyoxx57TDNnzuy0r8vlksvl8my3trZKktxut9xudyBT9tIxtjf7iHSRXKMzxoRuX4OM178DUbA1RuLadyWSj9muBHI8R8PxKoWnzkg7Lvrr8RqoUNXZ0/EOY0yPj5ojR47okksu0e9+9zvdcMMNnvann35ar732murr633G1NfXa9u2bcrKypLL5dIbb7yh0tJS1dTUaNq0aX6fp7i4WCUlJT7tFRUVSkhI6Ol0AQCARadPn9bChQvV0tKixMTETvsFdGWkg8Ph8No2xvi0dRg/frzGjx/v2c7JydHnn3+uVatWdRpGioqKVFBQ4NlubW1Venq6cnNzuyymO263W9XV1Zo1a5bi4uKC3k8ki+QaJxW/F7J9OQcZPZndrsd3DZKr3f+x198FW+PHxbPDOKvQi+RjtiuBHM/RcLxK4akz0o7n/nq8BipUdXa8stGdgMLIyJEjFRMTo+bmZq/2o0eP6uKLL+7xfqZMmaL169d3+rjT6ZTT6fRpj4uLC8nih2o/kSwSa3SdC/0fYVe7Iyz7jSSB1hhp695TkXjMdiWY4y4ajlcptHVG6jHR347XYPW2zp6ODegG1sGDBysrK0vV1dVe7dXV1V4v23Snrq5OqampgTw1AAAYoAJ+maagoECLFy9Wdna2cnJy9Morr6ixsVFLly6V9PVLLIcPH9brr78uSVq9erXGjh2rzMxMtbW1af369aqsrFRlZWVoKwEAAP1SwGHk9ttv1/Hjx/XjH/9YTU1NmjRpkjZv3qwxY8ZIkpqamrw+c6StrU2FhYU6fPiwhgwZoszMTFVVVWnOnDmhqwIAAPRbQd3A+qMf/Ug/+tGP/D5WXl7utb1s2TItW7YsmKcBAABRgO+mAQAAVhFGAACAVYQRAABgFWEEAABYFdQNrAAA9IWxy6tsT8GLM8bop9d9/Qm8nX2w22crv9fHs+r/uDICAACsIowAAACrCCMAAMAqwggAALCKG1j7oa5u6OrJzVUAAEQSrowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsIowAgAArCKMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrYm1PAACAgWTs8irbUwjYZyu/Z/X5uTICAACsivorI/0xwQIAMJBwZQQAAFhFGAEAAFYFFUZeeuklZWRkKD4+XllZWdq+fXuX/bdu3aqsrCzFx8fr0ksvVWlpaVCTBQAAA0/AYWTDhg16+OGH9eijj6qurk5Tp05VXl6eGhsb/fZvaGjQnDlzNHXqVNXV1WnFihV68MEHVVlZ2evJAwCA/i/gMPLcc8/p7rvv1j333KMJEyZo9erVSk9P15o1a/z2Ly0t1ejRo7V69WpNmDBB99xzj/7xH/9Rq1at6vXkAQBA/xfQu2na2tq0e/duLV++3Ks9NzdXtbW1fsfs2LFDubm5Xm2zZ89WWVmZ3G634uLifMa4XC65XC7PdktLiyTpr3/9q9xudyBT9uJ2u3X69GkdP37c87yxZ08Fvb9IFNtudPp0u2Ldg3Su3WF7OmETDXUGW+Px48fDOKvQ83de9geB/O2IhuNVio46B2qN5//dCNV5efLkSUmSMabLfgGFkWPHjuncuXO6+OKLvdovvvhiNTc3+x3T3Nzst//Zs2d17Ngxpaam+ox55plnVFJS4tOekZERyHSj1kLbE+gj0VBnMDWO/HnIp4EQiIbjVYqOOgdijeH+u3Hy5EklJSV1+nhQnzPicHinQWOMT1t3/f21dygqKlJBQYFnu729XX/96181YsSILp+nO62trUpPT9fnn3+uxMTEoPcTyaKhRik66oyGGqXoqDMaapSio85oqFEKXZ3GGJ08eVJpaWld9gsojIwcOVIxMTE+V0GOHj3qc/WjQ0pKit/+sbGxGjFihN8xTqdTTqfTq+3CCy8MZKpdSkxMHNAHkRQdNUrRUWc01ChFR53RUKMUHXVGQ41SaOrs6opIh4BuYB08eLCysrJUXV3t1V5dXa0bbrjB75icnByf/u+//76ys7P71evDAAAgPAJ+N01BQYFeffVVrV27VgcOHNAjjzyixsZGLV26VNLXL7EsWbLE03/p0qU6dOiQCgoKdODAAa1du1ZlZWUqLCwMXRUAAKDfCviekdtvv13Hjx/Xj3/8YzU1NWnSpEnavHmzxowZI0lqamry+syRjIwMbd68WY888ohefPFFpaWl6YUXXtCCBQtCV0UPOZ1OPfHEEz4vAQ0k0VCjFB11RkONUnTUGQ01StFRZzTUKPV9nQ7T3fttAAAAwojvpgEAAFYRRgAAgFWEEQAAYBVhBAAAWDWgwshTTz2lG264QQkJCT3+kDRjjIqLi5WWlqYhQ4ZoxowZ+tOf/uTVx+Vy6YEHHtDIkSM1dOhQff/739f//u//hqGC7n355ZdavHixkpKSlJSUpMWLF+vEiRNdjnE4HH5/fvazn3n6zJgxw+fxO+64I8zVdC6YOu+8806fGqZMmeLVpz+vpdvt1r/+67/qyiuv1NChQ5WWlqYlS5boyJEjXv1sr+VLL72kjIwMxcfHKysrS9u3b++y/9atW5WVlaX4+HhdeumlKi0t9elTWVmpiRMnyul0auLEidq4cWO4pt8jgdT49ttva9asWbrooouUmJionJwcvffee159ysvL/Z6jZ86cCXcpXQqkzpqaGr81HDx40KtfpK2lFFid/v7OOBwOZWZmevpE2npu27ZNc+fOVVpamhwOh955551ux/T5eWkGkH/7t38zzz33nCkoKDBJSUk9GrNy5UozbNgwU1lZafbt22duv/12k5qaalpbWz19li5dai655BJTXV1t9uzZY2bOnGmuvvpqc/bs2TBV0rlbbrnFTJo0ydTW1pra2lozadIkc+utt3Y5pqmpyetn7dq1xuFwmE8//dTTZ/r06ebee+/16nfixIlwl9OpYOrMz883t9xyi1cNx48f9+rTn9fyxIkT5uabbzYbNmwwBw8eNDt27DDXX3+9ycrK8upncy3ffPNNExcXZ375y1+a/fv3m4ceesgMHTrUHDp0yG////mf/zEJCQnmoYceMvv37ze//OUvTVxcnPnP//xPT5/a2loTExNjnn76aXPgwAHz9NNPm9jYWLNz584+qel8gdb40EMPmWeffdb8/ve/N3/+859NUVGRiYuLM3v27PH0WbdunUlMTPQ5V20KtM4PP/zQSDL19fVeNXzz3Iq0tTQm8DpPnDjhVd/nn39uhg8fbp544glPn0hbz82bN5tHH33UVFZWGklm48aNXfa3cV4OqDDSYd26dT0KI+3t7SYlJcWsXLnS03bmzBmTlJRkSktLjTFfH3hxcXHmzTff9PQ5fPiwGTRokNmyZUvI596V/fv3G0lei71jxw4jyRw8eLDH+5k3b5656aabvNqmT59uHnrooVBNtVeCrTM/P9/Mmzev08cH4lr+/ve/N5K8/nDaXMvrrrvOLF261KvtiiuuMMuXL/fbf9myZeaKK67wavunf/onM2XKFM/2bbfdZm655RavPrNnzzZ33HFHiGYdmEBr9GfixImmpKTEs93Tv1l9KdA6O8LIl19+2ek+I20tjen9em7cuNE4HA7z2WefedoicT079CSM2DgvB9TLNIFqaGhQc3OzcnNzPW1Op1PTp09XbW2tJGn37t1yu91efdLS0jRp0iRPn76yY8cOJSUl6frrr/e0TZkyRUlJST2eyxdffKGqqirdfffdPo/96le/0siRI5WZmanCwkLPVz/3td7UWVNTo+TkZI0bN0733nuvjh496nlsoK2lJLW0tMjhcPi8LGljLdva2rR7926v368k5ebmdlrTjh07fPrPnj1bu3btktvt7rJPX6+ZFFyN52tvb9fJkyc1fPhwr/avvvpKY8aM0ahRo3Trrbeqrq4uZPMOVG/qnDx5slJTU/Xd735XH374oddjkbSWUmjWs6ysTDfffLPngz87RNJ6BsrGeRnUt/YOFB1f4Hf+l/xdfPHFOnTokKfP4MGD9a1vfcunz/lfABhuzc3NSk5O9mlPTk7u8Vxee+01DRs2TD/4wQ+82hctWqSMjAylpKTo448/VlFRkf7whz/4fK9QXwi2zry8PP3whz/UmDFj1NDQoMcff1w33XSTdu/eLafTOeDW8syZM1q+fLkWLlzo9UVWttby2LFjOnfunN/zqbOampub/fY/e/asjh07ptTU1E779PWaScHVeL6f//znOnXqlG677TZP2xVXXKHy8nJdeeWVam1t1fPPP68bb7xRf/jDH3T55ZeHtIaeCKbO1NRUvfLKK8rKypLL5dIbb7yh7373u6qpqdG0adMkdb7eNtZS6v16NjU16d1331VFRYVXe6StZ6BsnJcRH0aKi4tVUlLSZZ+PPvpI2dnZQT+Hw+Hw2jbG+LSdryd9eqqnNUq+cw10LmvXrtWiRYsUHx/v1X7vvfd6/vekSZN0+eWXKzs7W3v27NE111zTo313J9x13n777Z7/PWnSJGVnZ2vMmDGqqqryCV+B7DcQfbWWbrdbd9xxh9rb2/XSSy95PdYXa9mVQM8nf/3Pbw/mHA2nYOfz61//WsXFxfrNb37jFUanTJnidbP1jTfeqGuuuUb/8R//oRdeeCF0Ew9QIHWOHz9e48eP92zn5OTo888/16pVqzxhJNB99pVg51ReXq4LL7xQ8+fP92qP1PUMRF+flxEfRu6///5u3wkwduzYoPadkpIi6esUmJqa6mk/evSoJ/GlpKSora1NX375pdf/oz569Gin31QcqJ7W+Mc//lFffPGFz2P/93//55NQ/dm+fbvq6+u1YcOGbvtec801iouL01/+8peQ/Qesr+rskJqaqjFjxugvf/mLpIGzlm63W7fddpsaGhr0wQcfdPv13uFYS39GjhypmJgYn/9n9M3z6XwpKSl++8fGxmrEiBFd9gnkWAiVYGrssGHDBt1999166623dPPNN3fZd9CgQbr22ms9x25f602d3zRlyhStX7/esx1Jayn1rk5jjNauXavFixdr8ODBXfa1vZ6BsnJeBnWnSYQL9AbWZ5991tPmcrn83sC6YcMGT58jR45Yvenxv//7vz1tO3fu7PFNj/n5+T7vvOjMvn37jCSzdevWoOcbrN7W2eHYsWPG6XSa1157zRgzMNayra3NzJ8/32RmZpqjR4/26Ln6ci2vu+4688///M9ebRMmTOjyBtYJEyZ4tS1dutTnRrm8vDyvPrfccovVG1gDqdEYYyoqKkx8fHy3Nw52aG9vN9nZ2eauu+7qzVR7JZg6z7dgwQIzc+ZMz3akraUxwdfZccPuvn37un2OSFjPDurhDax9fV4OqDBy6NAhU1dXZ0pKSswFF1xg6urqTF1dnTl58qSnz/jx483bb7/t2V65cqVJSkoyb7/9ttm3b5/5+7//e79v7R01apT5r//6L7Nnzx5z0003WX076FVXXWV27NhhduzYYa688kqft4OeX6MxxrS0tJiEhASzZs0an31+8sknpqSkxHz00UemoaHBVFVVmSuuuMJMnjzZSo3GBF7nyZMnzb/8y7+Y2tpa09DQYD788EOTk5NjLrnkkgGzlm6323z/+983o0aNMnv37vV6y6DL5TLG2F/LjrdJlpWVmf3795uHH37YDB061PNOg+XLl5vFixd7+ne8hfCRRx4x+/fvN2VlZT5vIfzd735nYmJizMqVK82BAwfMypUrI+KtvT2tsaKiwsTGxpoXX3yx07dbFxcXmy1btphPP/3U1NXVmbvuusvExsZ6hdW+Fmid//7v/242btxo/vznP5uPP/7YLF++3EgylZWVnj6RtpbGBF5nh3/4h38w119/vd99Rtp6njx50vPfQ0nmueeeM3V1dZ534UXCeTmgwkh+fr6R5PPz4YcfevpIMuvWrfNst7e3myeeeMKkpKQYp9Nppk2b5pN0//a3v5n777/fDB8+3AwZMsTceuutprGxsY+q8nb8+HGzaNEiM2zYMDNs2DCzaNEin7fSnV+jMca8/PLLZsiQIX4/b6KxsdFMmzbNDB8+3AwePNhcdtll5sEHH/T5jI6+FGidp0+fNrm5ueaiiy4ycXFxZvTo0SY/P99nnfrzWjY0NPg9vr95jEfCWr744otmzJgxZvDgweaaa67xuiKTn59vpk+f7tW/pqbGTJ482QwePNiMHTvWb2B+6623zPjx401cXJy54oorvP4DZ0MgNU6fPt3vmuXn53v6PPzww2b06NFm8ODB5qKLLjK5ubmmtra2DyvyL5A6n332WXPZZZeZ+Ph4861vfct85zvfMVVVVT77jLS1NCbwY/bEiRNmyJAh5pVXXvG7v0hbz46rOJ0dg5FwXjqM+f93pQAAAFgQ1Z8zAgAA7COMAAAAqwgjAADAKsIIAACwijACAACsIowAAACrCCMAAMAqwggAALCKMAIAAKwijAAAAKsIIwAAwCrCCAAAsOr/AeRY1qBLbUySAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "himorna_peaks[\"corr\"].hist()# данные уже отфильтрованы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "ed005b42-caf3-4e8c-95b2-a254d7429025",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexing_himorna(hm, lnc_ens, subdf):\n",
    "    to_index = HIMORNA_INDEX / hm\n",
    "    to_index.mkdir(exist_ok=True, parents=True)\n",
    "    to_save = to_index / f\"{hm}-{lnc_ens}-.tsv.gz\"\n",
    "    subdf.to_csv(to_save, sep=\"\\t\", compression=\"gzip\", index=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "803324ec-717d-494c-903f-045c8ec30a77",
   "metadata": {},
   "outputs": [],
   "source": [
    "with Parallel(n_jobs=-1) as pool:\n",
    "    chunks = pool(\n",
    "        delayed(indexing_himorna)(hm, lnc_ens, subdf)\n",
    "        for (hm, lnc_ens), subdf in himorna_peaks.groupby(by=[\"hm\", \"lnc_ens\"])\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2370a0a-f2a1-47c0-988f-9b849b8ced61",
   "metadata": {},
   "source": [
    "## fantom ips deg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a8c9a06-330c-4132-8b39-57f2d58bd9e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DEG.geneID</th>\n",
       "      <th>DEG.HGNC</th>\n",
       "      <th>EdgeR.log2FC</th>\n",
       "      <th>EdgeR.log2CPM</th>\n",
       "      <th>EdgeR.pvalue</th>\n",
       "      <th>EdgeR.FDR</th>\n",
       "      <th>EdgeR.Zscore</th>\n",
       "      <th>target.gapmerID</th>\n",
       "      <th>target.geneID</th>\n",
       "      <th>target.HGNC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATG00000000155</td>\n",
       "      <td>CATG00000000155.1</td>\n",
       "      <td>0.257141</td>\n",
       "      <td>3.591144</td>\n",
       "      <td>0.550588</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.620927</td>\n",
       "      <td>G0234787_01</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CATG00000001164</td>\n",
       "      <td>CATG00000001164.1</td>\n",
       "      <td>0.706131</td>\n",
       "      <td>3.805757</td>\n",
       "      <td>0.044323</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.757448</td>\n",
       "      <td>G0234787_01</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CATG00000001318</td>\n",
       "      <td>CATG00000001318.1</td>\n",
       "      <td>-0.074655</td>\n",
       "      <td>4.766566</td>\n",
       "      <td>0.913694</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.052701</td>\n",
       "      <td>G0234787_01</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CATG00000002817</td>\n",
       "      <td>CATG00000002817.1</td>\n",
       "      <td>-0.201767</td>\n",
       "      <td>3.841792</td>\n",
       "      <td>0.716487</td>\n",
       "      <td>1.0</td>\n",
       "      <td>-0.414103</td>\n",
       "      <td>G0234787_01</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CATG00000003248</td>\n",
       "      <td>CATG00000003248.1</td>\n",
       "      <td>-0.155609</td>\n",
       "      <td>4.592817</td>\n",
       "      <td>0.650460</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.026003</td>\n",
       "      <td>G0234787_01</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        DEG.geneID           DEG.HGNC  EdgeR.log2FC  EdgeR.log2CPM  \\\n",
       "0  CATG00000000155  CATG00000000155.1      0.257141       3.591144   \n",
       "1  CATG00000001164  CATG00000001164.1      0.706131       3.805757   \n",
       "2  CATG00000001318  CATG00000001318.1     -0.074655       4.766566   \n",
       "3  CATG00000002817  CATG00000002817.1     -0.201767       3.841792   \n",
       "4  CATG00000003248  CATG00000003248.1     -0.155609       4.592817   \n",
       "\n",
       "   EdgeR.pvalue  EdgeR.FDR  EdgeR.Zscore target.gapmerID    target.geneID  \\\n",
       "0      0.550588        1.0      0.620927     G0234787_01  ENSG00000234787   \n",
       "1      0.044323        1.0      1.757448     G0234787_01  ENSG00000234787   \n",
       "2      0.913694        1.0      0.052701     G0234787_01  ENSG00000234787   \n",
       "3      0.716487        1.0     -0.414103     G0234787_01  ENSG00000234787   \n",
       "4      0.650460        1.0      0.026003     G0234787_01  ENSG00000234787   \n",
       "\n",
       "  target.HGNC  \n",
       "0   LINC00458  \n",
       "1   LINC00458  \n",
       "2   LINC00458  \n",
       "3   LINC00458  \n",
       "4   LINC00458  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_ips = IPS / \"EdgeR.table.all.target.tsv.gz\"\n",
    "ips_expr = pd.read_csv(to_ips, sep=\"\\t\")\n",
    "ips_expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d713b1c-3888-4bbc-aea7-5cecbd574fb6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3445156, 10)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ips_expr.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0a850eeb-e3fd-4629-9698-8af581cbb183",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DEG.geneID</th>\n",
       "      <th>DEG.HGNC</th>\n",
       "      <th>EdgeR.log2FC</th>\n",
       "      <th>EdgeR.log2CPM</th>\n",
       "      <th>EdgeR.pvalue</th>\n",
       "      <th>EdgeR.FDR</th>\n",
       "      <th>EdgeR.Zscore</th>\n",
       "      <th>target.gapmerID</th>\n",
       "      <th>target.geneID</th>\n",
       "      <th>target.HGNC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000205531</td>\n",
       "      <td>NAP1L4</td>\n",
       "      <td>-0.691972</td>\n",
       "      <td>7.349218</td>\n",
       "      <td>3.899614e-06</td>\n",
       "      <td>0.042455</td>\n",
       "      <td>-5.061144</td>\n",
       "      <td>G0234787_08</td>\n",
       "      <td>ENSG00000234787</td>\n",
       "      <td>LINC00458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000089916</td>\n",
       "      <td>GPATCH2L</td>\n",
       "      <td>-1.041246</td>\n",
       "      <td>5.600935</td>\n",
       "      <td>3.913478e-05</td>\n",
       "      <td>0.043336</td>\n",
       "      <td>-3.953852</td>\n",
       "      <td>R0057746_03</td>\n",
       "      <td>CATG00000070943</td>\n",
       "      <td>CATG00000070943.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000099194</td>\n",
       "      <td>SCD</td>\n",
       "      <td>-0.635000</td>\n",
       "      <td>8.662415</td>\n",
       "      <td>2.054724e-07</td>\n",
       "      <td>0.000870</td>\n",
       "      <td>-2.709444</td>\n",
       "      <td>R0057746_03</td>\n",
       "      <td>CATG00000070943</td>\n",
       "      <td>CATG00000070943.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000113161</td>\n",
       "      <td>HMGCR</td>\n",
       "      <td>-0.912822</td>\n",
       "      <td>6.861125</td>\n",
       "      <td>1.601495e-05</td>\n",
       "      <td>0.029059</td>\n",
       "      <td>-3.682696</td>\n",
       "      <td>R0057746_03</td>\n",
       "      <td>CATG00000070943</td>\n",
       "      <td>CATG00000070943.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000128829</td>\n",
       "      <td>EIF2AK4</td>\n",
       "      <td>-0.887877</td>\n",
       "      <td>6.605389</td>\n",
       "      <td>2.397361e-07</td>\n",
       "      <td>0.000870</td>\n",
       "      <td>-4.709886</td>\n",
       "      <td>R0057746_03</td>\n",
       "      <td>CATG00000070943</td>\n",
       "      <td>CATG00000070943.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        DEG.geneID  DEG.HGNC  EdgeR.log2FC  EdgeR.log2CPM  EdgeR.pvalue  \\\n",
       "0  ENSG00000205531    NAP1L4     -0.691972       7.349218  3.899614e-06   \n",
       "1  ENSG00000089916  GPATCH2L     -1.041246       5.600935  3.913478e-05   \n",
       "2  ENSG00000099194       SCD     -0.635000       8.662415  2.054724e-07   \n",
       "3  ENSG00000113161     HMGCR     -0.912822       6.861125  1.601495e-05   \n",
       "4  ENSG00000128829   EIF2AK4     -0.887877       6.605389  2.397361e-07   \n",
       "\n",
       "   EdgeR.FDR  EdgeR.Zscore target.gapmerID    target.geneID        target.HGNC  \n",
       "0   0.042455     -5.061144     G0234787_08  ENSG00000234787          LINC00458  \n",
       "1   0.043336     -3.953852     R0057746_03  CATG00000070943  CATG00000070943.1  \n",
       "2   0.000870     -2.709444     R0057746_03  CATG00000070943  CATG00000070943.1  \n",
       "3   0.029059     -3.682696     R0057746_03  CATG00000070943  CATG00000070943.1  \n",
       "4   0.000870     -4.709886     R0057746_03  CATG00000070943  CATG00000070943.1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ips_expr_filt = ips_expr[(((ips_expr[\"EdgeR.log2FC\"] < 0.5) & (ips_expr[\"EdgeR.Zscore\"] < -1.64)) | \n",
    "                                ((ips_expr[\"EdgeR.log2FC\"] > 0.5) & (ips_expr[\"EdgeR.Zscore\"] > 1.64))) & \n",
    "                                (ips_expr[\"EdgeR.FDR\"] < 0.05)]\n",
    "ips_expr_filt.reset_index(inplace=True, drop=True)\n",
    "ips_expr_filt.head()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "633d6207-724e-4d33-acd1-aa2222dde785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16294, 10)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ips_expr_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b788a7ec-3f8e-4514-aacf-87142ce5ccaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key, subdf in ips_expr_filt.groupby(by=[\"target.geneID\", \"target.gapmerID\"]):\n",
    "    print(key, subdf.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "172b1dca-d38c-4866-ba75-6ee2bd1d3f91",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexing_deg(lnc_ens, aso, subdf):\n",
    "    to_index = IPS_INDEX / lnc_ens\n",
    "    to_index.mkdir(exist_ok=True, parents=True)\n",
    "    to_save = to_index / f\"{lnc_ens}-{aso}-.tsv.gz\"\n",
    "    subdf.to_csv(to_save, sep=\"\\t\", compression=\"gzip\", index=False)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "4483b6d6-7af0-4728-aa8d-699279e31fd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "with Parallel(n_jobs=-1) as pool:\n",
    "    chunks = pool(\n",
    "        delayed(indexing_deg)(lnc_ens, aso, subdf)\n",
    "        for (lnc_ens, aso), subdf in ips_expr_filt.groupby(by=[\"target.geneID\", \"target.gapmerID\"])\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47875693-4f85-42ea-887f-53d84c4795ba",
   "metadata": {},
   "source": [
    "## fantom cat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "acd3ebeb-9b5f-4645-bbec-8b0292f06ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fantom_lv2 = pd.read_csv(FANTOM / \"FANTOM_CAT.lv2_permissive.info_table.gene.tsv.gz\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "a1075e23-1401-4e4f-ada8-478c035ac0b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fantom_lv2_mapping = fantom_lv2[[\"geneID_without_v\", \"geneName\"]].set_index(\"geneID_without_v\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e469a0f-353f-4983-b198-a64f2b38e470",
   "metadata": {},
   "source": [
    "## matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "0242087a-b704-4100-87be-a58ea4bcd6bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_matrix(to_deg, to_himorna):\n",
    "    deg_gene_ids = np.asarray(pd.read_csv(to_deg, sep=\"\\t\", usecols=[\"DEG.geneID\"])[\"DEG.geneID\"])\n",
    "    corr_gene_ids = np.asarray(pd.read_csv(to_himorna, sep=\"\\t\", usecols=[\"gene_ens\"])[\"gene_ens\"])\n",
    "    \n",
    "    deg_mask = fantom_lv2[\"geneID_without_v\"].isin(deg_gene_ids)\n",
    "    corr_mask = fantom_lv2[\"geneID_without_v\"].isin(corr_gene_ids)\n",
    "    TP = deg_mask & corr_mask\n",
    "    TN = deg_mask & ~corr_mask\n",
    "    FP = ~deg_mask & corr_mask\n",
    "    FN = ~(deg_mask | corr_mask)\n",
    "    \n",
    "    to_save = MATRICES / to_himorna.name.split('-')[0] / f\"{to_himorna.name.split('-')[1]}-{to_deg.name.split('-')[1]}-.npy\"\n",
    "    if to_save.exists():\n",
    "        print(f\"Skipped: ({to_himorna.name.split('-')[0]}) {to_himorna.name.split('-')[1]} {to_deg.name.split('-')[1]}\")\n",
    "        return\n",
    "    to_save.parent.mkdir(exist_ok=True, parents=True)\n",
    "    np.save(to_save, np.stack((TP, TN, FP, FN)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "ff612a3b-69cf-4ed2-8f8e-bb31a218aa07",
   "metadata": {},
   "outputs": [],
   "source": [
    "with Parallel(n_jobs=-1) as pool:\n",
    "    chunks = pool(\n",
    "        delayed(create_matrix)(to_deg, to_himorna)\n",
    "        for to_deg in IPS_INDEX.glob(('*/*.tsv.gz')) for to_himorna in HIMORNA_INDEX.glob((f'*/*-{to_deg.name.split(\"-\")[0]}-.tsv.gz'))\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d04f000-0b61-4bb5-85aa-301631a3bcfd",
   "metadata": {},
   "source": [
    "## fisher exact test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "id": "fb215de3-7691-4962-a334-67f304a3e757",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fisher_exact_test(to_matrice):\n",
    "    matrice = np.sum(np.load(to_matrice), axis=1)\n",
    "    fisher_exact_res = fisher_exact([[matrice[0], matrice[1]], \n",
    "                         [matrice[2], matrice[3]]], alternative='two-sided')\n",
    "    return fisher_exact_res[1]\n",
    "    #return matrice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "1e6b20c6-fa6d-4273-ba2c-8eefe86dafaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def job(hm):\n",
    "    p_values = []\n",
    "    lncrna = []\n",
    "    aso = []\n",
    "    for to_matrice in (MATRICES / hm).glob('*'):\n",
    "        p_values.append(fisher_exact_test(to_matrice))\n",
    "        lncrna.append(to_matrice.name.split(\"-\")[0])\n",
    "        aso.append(to_matrice.name.split(\"-\")[1])\n",
    "    df = pd.DataFrame({\"hm\": [hm for i in range(len(p_values))],\n",
    "                       \"lnc_ens\": lncrna,\n",
    "                       \"aso\": aso,\n",
    "                       \"p_value\": p_values})\n",
    "    p_value_adj_bonf = multipletests(df[\"p_value\"], method='bonferroni', is_sorted=False, returnsorted=False)[1]\n",
    "    df['p_value_adj_bonf'] = p_value_adj_bonf\n",
    "    to_save = FISHER / f\"{hm}.tsv\"\n",
    "    if to_save.exists():\n",
    "        print(f\"Skipped: {hm}\")\n",
    "        return\n",
    "    to_save.parent.mkdir(exist_ok=True, parents=True)\n",
    "    df.to_csv(to_save, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "a869e112-170b-4cff-a040-ba62e2c2dbe6",
   "metadata": {},
   "outputs": [],
   "source": [
    "with Parallel(n_jobs=-1) as pool:\n",
    "    chunks = pool(\n",
    "        delayed(job)(hm.name)\n",
    "        for hm in MATRICES.iterdir())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "324a823a-03ce-4bc0-8a78-3e9db5fc9590",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "efb25050-b0b0-4371-8f4f-60b0576068e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fisher_results = {to_df.stem: pd.read_csv(to_df ,sep=\"\\t\") for to_df in FISHER.glob(\"*.tsv\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "id": "cb3c2f6d-d16d-479b-92fc-a7ff5bd183f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "treshold_p_value = 0.05\n",
    "fisher_results_filt = []\n",
    "for hm in fisher_results:\n",
    "    fisher_results_filt.append(fisher_results[hm].query(f\"p_value_adj_bonf < {treshold_p_value}\").reset_index(drop=True))\n",
    "fisher_results_filt = pd.concat(fisher_results_filt).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "851dd601-212a-4057-9d81-17e0db075687",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fisher_results_filt[\"lnc\"] = fisher_results_filt[\"lnc_ens\"].apply(lambda x: fantom_lv2_mapping.loc[x, \"geneName\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "517bdcf2-f6c9-4554-a2e7-48961d066b82",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hm</th>\n",
       "      <th>lnc_ens</th>\n",
       "      <th>aso</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj_bonf</th>\n",
       "      <th>lnc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000235560</td>\n",
       "      <td>G0235560_01</td>\n",
       "      <td>2.803825e-10</td>\n",
       "      <td>1.878562e-08</td>\n",
       "      <td>AC002310.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000235560</td>\n",
       "      <td>G0235560_03</td>\n",
       "      <td>3.290078e-06</td>\n",
       "      <td>2.204352e-04</td>\n",
       "      <td>AC002310.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000227403</td>\n",
       "      <td>G0227403_04</td>\n",
       "      <td>1.793569e-04</td>\n",
       "      <td>1.381048e-02</td>\n",
       "      <td>AC009299.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>ENSG00000227403</td>\n",
       "      <td>G0227403_04</td>\n",
       "      <td>1.557503e-04</td>\n",
       "      <td>1.386178e-02</td>\n",
       "      <td>AC009299.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000212978</td>\n",
       "      <td>G0212978_03</td>\n",
       "      <td>1.335501e-04</td>\n",
       "      <td>1.028335e-02</td>\n",
       "      <td>AC016747.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000212978</td>\n",
       "      <td>G0212978_03</td>\n",
       "      <td>2.382738e-04</td>\n",
       "      <td>2.430393e-02</td>\n",
       "      <td>AC016747.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000225975</td>\n",
       "      <td>G0225975_05</td>\n",
       "      <td>2.575401e-04</td>\n",
       "      <td>2.626909e-02</td>\n",
       "      <td>AC074138.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>H3K9ac</td>\n",
       "      <td>ENSG00000245937</td>\n",
       "      <td>G0245937_01</td>\n",
       "      <td>2.845680e-04</td>\n",
       "      <td>1.622037e-02</td>\n",
       "      <td>CTC-228N24.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000245937</td>\n",
       "      <td>G0245937_01</td>\n",
       "      <td>1.844638e-04</td>\n",
       "      <td>1.881531e-02</td>\n",
       "      <td>CTC-228N24.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000249042</td>\n",
       "      <td>G0249042_08</td>\n",
       "      <td>4.591768e-08</td>\n",
       "      <td>3.076485e-06</td>\n",
       "      <td>CTD-2015H6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000249042</td>\n",
       "      <td>G0249042_08</td>\n",
       "      <td>3.613109e-06</td>\n",
       "      <td>2.782094e-04</td>\n",
       "      <td>CTD-2015H6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>ENSG00000249042</td>\n",
       "      <td>G0249042_08</td>\n",
       "      <td>1.594663e-04</td>\n",
       "      <td>1.419250e-02</td>\n",
       "      <td>CTD-2015H6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000196668</td>\n",
       "      <td>G0196668_08</td>\n",
       "      <td>1.764837e-15</td>\n",
       "      <td>1.358924e-13</td>\n",
       "      <td>LINC00173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000196668</td>\n",
       "      <td>G0196668_08</td>\n",
       "      <td>1.084557e-12</td>\n",
       "      <td>7.266531e-11</td>\n",
       "      <td>LINC00173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000196668</td>\n",
       "      <td>G0196668_02</td>\n",
       "      <td>1.681718e-12</td>\n",
       "      <td>1.294923e-10</td>\n",
       "      <td>LINC00173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000196668</td>\n",
       "      <td>G0196668_02</td>\n",
       "      <td>2.032593e-09</td>\n",
       "      <td>1.361838e-07</td>\n",
       "      <td>LINC00173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>H3K9ac</td>\n",
       "      <td>ENSG00000196668</td>\n",
       "      <td>G0196668_08</td>\n",
       "      <td>2.121320e-04</td>\n",
       "      <td>1.209152e-02</td>\n",
       "      <td>LINC00173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>ENSG00000250337</td>\n",
       "      <td>G0250337_05</td>\n",
       "      <td>1.044787e-04</td>\n",
       "      <td>8.044861e-03</td>\n",
       "      <td>LINC01021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000250337</td>\n",
       "      <td>G0250337_05</td>\n",
       "      <td>2.983490e-04</td>\n",
       "      <td>1.998938e-02</td>\n",
       "      <td>LINC01021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000234608</td>\n",
       "      <td>G0234608_03</td>\n",
       "      <td>1.286297e-17</td>\n",
       "      <td>1.312023e-15</td>\n",
       "      <td>MAPKAPK5-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000234608</td>\n",
       "      <td>G0234608_AD_08</td>\n",
       "      <td>2.163791e-08</td>\n",
       "      <td>2.207067e-06</td>\n",
       "      <td>MAPKAPK5-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>H3K4me3</td>\n",
       "      <td>ENSG00000247556</td>\n",
       "      <td>G0247556_09</td>\n",
       "      <td>8.580857e-23</td>\n",
       "      <td>8.752474e-21</td>\n",
       "      <td>OIP5-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>ENSG00000237976</td>\n",
       "      <td>G0237976_03</td>\n",
       "      <td>9.157260e-05</td>\n",
       "      <td>8.149962e-03</td>\n",
       "      <td>RP11-126K1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>ENSG00000233184</td>\n",
       "      <td>G0233184_04</td>\n",
       "      <td>3.061645e-24</td>\n",
       "      <td>2.143151e-22</td>\n",
       "      <td>RP11-421L21.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H3K36me3</td>\n",
       "      <td>ENSG00000233184</td>\n",
       "      <td>G0233184_03</td>\n",
       "      <td>1.973102e-07</td>\n",
       "      <td>1.381171e-05</td>\n",
       "      <td>RP11-421L21.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>ENSG00000230623</td>\n",
       "      <td>G0230623_09</td>\n",
       "      <td>9.675634e-06</td>\n",
       "      <td>8.611314e-04</td>\n",
       "      <td>RP11-469A15.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>H3K4me1</td>\n",
       "      <td>ENSG00000263146</td>\n",
       "      <td>G0263146_04</td>\n",
       "      <td>8.329297e-13</td>\n",
       "      <td>7.413075e-11</td>\n",
       "      <td>RP11-849I19.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000263146</td>\n",
       "      <td>G0263146_04</td>\n",
       "      <td>2.895048e-11</td>\n",
       "      <td>2.229187e-09</td>\n",
       "      <td>RP11-849I19.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H3K27me3</td>\n",
       "      <td>ENSG00000234546</td>\n",
       "      <td>G0234546_02</td>\n",
       "      <td>1.625129e-08</td>\n",
       "      <td>1.251349e-06</td>\n",
       "      <td>RP3-510D11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000234546</td>\n",
       "      <td>G0234546_02</td>\n",
       "      <td>5.593146e-07</td>\n",
       "      <td>3.747408e-05</td>\n",
       "      <td>RP3-510D11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>H3K27ac</td>\n",
       "      <td>ENSG00000234546</td>\n",
       "      <td>G0234546_02</td>\n",
       "      <td>5.314236e-06</td>\n",
       "      <td>4.091962e-04</td>\n",
       "      <td>RP3-510D11.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>H3K4me2</td>\n",
       "      <td>ENSG00000246273</td>\n",
       "      <td>G0246273_02</td>\n",
       "      <td>1.078811e-04</td>\n",
       "      <td>7.228035e-03</td>\n",
       "      <td>SBF2-AS1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          hm          lnc_ens             aso       p_value  p_value_adj_bonf  \\\n",
       "5    H3K4me2  ENSG00000235560     G0235560_01  2.803825e-10      1.878562e-08   \n",
       "8    H3K4me2  ENSG00000235560     G0235560_03  3.290078e-06      2.204352e-04   \n",
       "20   H3K27ac  ENSG00000227403     G0227403_04  1.793569e-04      1.381048e-02   \n",
       "30   H3K4me1  ENSG00000227403     G0227403_04  1.557503e-04      1.386178e-02   \n",
       "23   H3K27ac  ENSG00000212978     G0212978_03  1.335501e-04      1.028335e-02   \n",
       "16   H3K4me3  ENSG00000212978     G0212978_03  2.382738e-04      2.430393e-02   \n",
       "13   H3K4me3  ENSG00000225975     G0225975_05  2.575401e-04      2.626909e-02   \n",
       "26    H3K9ac  ENSG00000245937     G0245937_01  2.845680e-04      1.622037e-02   \n",
       "17   H3K4me3  ENSG00000245937     G0245937_01  1.844638e-04      1.881531e-02   \n",
       "10   H3K4me2  ENSG00000249042     G0249042_08  4.591768e-08      3.076485e-06   \n",
       "22   H3K27ac  ENSG00000249042     G0249042_08  3.613109e-06      2.782094e-04   \n",
       "31   H3K4me1  ENSG00000249042     G0249042_08  1.594663e-04      1.419250e-02   \n",
       "21   H3K27ac  ENSG00000196668     G0196668_08  1.764837e-15      1.358924e-13   \n",
       "9    H3K4me2  ENSG00000196668     G0196668_08  1.084557e-12      7.266531e-11   \n",
       "19   H3K27ac  ENSG00000196668     G0196668_02  1.681718e-12      1.294923e-10   \n",
       "4    H3K4me2  ENSG00000196668     G0196668_02  2.032593e-09      1.361838e-07   \n",
       "25    H3K9ac  ENSG00000196668     G0196668_08  2.121320e-04      1.209152e-02   \n",
       "0   H3K27me3  ENSG00000250337     G0250337_05  1.044787e-04      8.044861e-03   \n",
       "6    H3K4me2  ENSG00000250337     G0250337_05  2.983490e-04      1.998938e-02   \n",
       "12   H3K4me3  ENSG00000234608     G0234608_03  1.286297e-17      1.312023e-15   \n",
       "14   H3K4me3  ENSG00000234608  G0234608_AD_08  2.163791e-08      2.207067e-06   \n",
       "15   H3K4me3  ENSG00000247556     G0247556_09  8.580857e-23      8.752474e-21   \n",
       "29   H3K4me1  ENSG00000237976     G0237976_03  9.157260e-05      8.149962e-03   \n",
       "3   H3K36me3  ENSG00000233184     G0233184_04  3.061645e-24      2.143151e-22   \n",
       "2   H3K36me3  ENSG00000233184     G0233184_03  1.973102e-07      1.381171e-05   \n",
       "27   H3K4me1  ENSG00000230623     G0230623_09  9.675634e-06      8.611314e-04   \n",
       "28   H3K4me1  ENSG00000263146     G0263146_04  8.329297e-13      7.413075e-11   \n",
       "18   H3K27ac  ENSG00000263146     G0263146_04  2.895048e-11      2.229187e-09   \n",
       "1   H3K27me3  ENSG00000234546     G0234546_02  1.625129e-08      1.251349e-06   \n",
       "11   H3K4me2  ENSG00000234546     G0234546_02  5.593146e-07      3.747408e-05   \n",
       "24   H3K27ac  ENSG00000234546     G0234546_02  5.314236e-06      4.091962e-04   \n",
       "7    H3K4me2  ENSG00000246273     G0246273_02  1.078811e-04      7.228035e-03   \n",
       "\n",
       "              lnc  \n",
       "5     AC002310.12  \n",
       "8     AC002310.12  \n",
       "20     AC009299.3  \n",
       "30     AC009299.3  \n",
       "23     AC016747.3  \n",
       "16     AC016747.3  \n",
       "13     AC074138.3  \n",
       "26   CTC-228N24.3  \n",
       "17   CTC-228N24.3  \n",
       "10   CTD-2015H6.3  \n",
       "22   CTD-2015H6.3  \n",
       "31   CTD-2015H6.3  \n",
       "21      LINC00173  \n",
       "9       LINC00173  \n",
       "19      LINC00173  \n",
       "4       LINC00173  \n",
       "25      LINC00173  \n",
       "0       LINC01021  \n",
       "6       LINC01021  \n",
       "12   MAPKAPK5-AS1  \n",
       "14   MAPKAPK5-AS1  \n",
       "15       OIP5-AS1  \n",
       "29   RP11-126K1.6  \n",
       "3   RP11-421L21.3  \n",
       "2   RP11-421L21.3  \n",
       "27  RP11-469A15.2  \n",
       "28  RP11-849I19.1  \n",
       "18  RP11-849I19.1  \n",
       "1    RP3-510D11.2  \n",
       "11   RP3-510D11.2  \n",
       "24   RP3-510D11.2  \n",
       "7        SBF2-AS1  "
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fisher_results_filt.sort_values(by=[\"lnc\", \"p_value_adj_bonf\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "dcd2c7e5-c579-459b-9dfe-02a389e6663b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['LINC01021', 'RP3-510D11.2', 'RP11-421L21.3', 'LINC00173',\n",
       "       'AC002310.12', 'SBF2-AS1', 'CTD-2015H6.3', 'MAPKAPK5-AS1',\n",
       "       'AC074138.3', 'OIP5-AS1', 'AC016747.3', 'CTC-228N24.3',\n",
       "       'RP11-849I19.1', 'AC009299.3', 'RP11-469A15.2', 'RP11-126K1.6'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fisher_results_filt[\"lnc\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87fb1a4c-a8ec-44bd-8de3-1840e5d0d5af",
   "metadata": {},
   "source": [
    "## Pubmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "id": "89ceaf40-0a4d-4418-abb4-5bd12fbc3fd2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PubMed ID: 36342127\n",
      "PubMed ID: 35173309\n",
      "PubMed ID: 35012438\n",
      "PubMed ID: 34469433\n",
      "PubMed ID: 34026327\n",
      "PubMed ID: 32774710\n",
      "PubMed ID: 32774475\n",
      "PubMed ID: 29262524\n",
      "PubMed ID: 28877474\n",
      "PubMed ID: 26183718\n",
      "PubMed ID: 38182665\n",
      "PubMed ID: 38069335\n",
      "PubMed ID: 37365917\n",
      "PubMed ID: 37258811\n",
      "PubMed ID: 37081063\n",
      "PubMed ID: 36894691\n",
      "PubMed ID: 36856154\n",
      "PubMed ID: 36774758\n",
      "PubMed ID: 36627670\n",
      "PubMed ID: 36606322\n",
      "PubMed ID: 38279439\n",
      "PubMed ID: 37593490\n",
      "PubMed ID: 37538086\n",
      "PubMed ID: 37445702\n",
      "PubMed ID: 37286777\n",
      "PubMed ID: 37131869\n",
      "PubMed ID: 37072750\n",
      "PubMed ID: 36927189\n",
      "PubMed ID: 36819921\n",
      "PubMed ID: 36754932\n",
      "PubMed ID: 38225924\n",
      "PubMed ID: 37154079\n",
      "PubMed ID: 36867352\n",
      "PubMed ID: 36567178\n",
      "PubMed ID: 36510238\n",
      "PubMed ID: 36445573\n",
      "PubMed ID: 36379256\n",
      "PubMed ID: 36009801\n",
      "PubMed ID: 35674862\n",
      "PubMed ID: 35468721\n",
      "PubMed ID: 38495196\n",
      "PubMed ID: 38461631\n",
      "PubMed ID: 38441253\n",
      "PubMed ID: 38419565\n",
      "PubMed ID: 38402377\n",
      "PubMed ID: 38398086\n",
      "PubMed ID: 38326441\n",
      "PubMed ID: 38294019\n",
      "PubMed ID: 38248871\n",
      "PubMed ID: 38194746\n",
      "PubMed ID: 34735975\n",
      "PubMed ID: 36387469\n"
     ]
    }
   ],
   "source": [
    "# Установка адреса электронной почты, необходимого для использования API PubMed\n",
    "Entrez.email = \"m.agletdinov2018@gmail.com\"\n",
    "\n",
    "def search_pubmed(query):\n",
    "    # Поиск статей по запросу\n",
    "    handle = Entrez.esearch(db=\"pubmed\", term=query, retmax=10)\n",
    "    record = Entrez.read(handle)\n",
    "    handle.close()\n",
    "    return record\n",
    "\n",
    "def fetch_abstract(pmid):\n",
    "    # Получение абстракта статьи по ее PubMed ID (PMID)\n",
    "    handle = Entrez.efetch(db=\"pubmed\", id=pmid, rettype=\"abstract\", retmode=\"text\")\n",
    "    abstract = handle.read()\n",
    "    handle.close()\n",
    "    return abstract.replace(\"\\n\",\"\")\n",
    "\n",
    "pubmed_res = defaultdict(list)\n",
    "for query in fisher_results_filt[\"lnc\"].unique():\n",
    "    print(f\"{query)\\n\")\n",
    "    record = search_pubmed(query)\n",
    "    for pmid in record[\"IdList\"]:\n",
    "        abstract = fetch_abstract(pmid)\n",
    "        print(\"\\tPubMed ID:\", pmid)\n",
    "        pubmed_res[query].append([pmid, abstract])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "id": "8bd28bc0-5e3d-4d99-a325-ad03745b787a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['38182665',\n",
       "  \"1. Commun Biol. 2024 Jan 5;7(1):35. doi: 10.1038/s42003-023-05725-x.Differential methylation analysis in neuropathologically confirmed dementia with Lewy bodies.Reho P(1)(2), Saez-Atienzar S(3), Ruffo P(3)(4), Solaiman S(1), Shah Z(1), Chia R(3), Kaivola K(1), Traynor BJ(3), Tilley BS(5), Gentleman SM(5), Hodges AK(6), Aarsland D(6)(7), Monuki ES(8), Newell KL(9), Woltjer R(10), Albert MS(11), Dawson TM(11)(12)(13)(14), Rosenthal LS(11), Troncoso JC(15), Pletnikova O(16), Serrano GE(17), Beach TG(17), Easwaran HP(18), Scholz SW(19)(20).Author information:(1)Neurodegenerative Diseases Research Unit, National Institute of Neurological Disorders and Stroke, Bethesda, MD, USA.(2)Laboratory of Precision Environmental Health, Department of Environmental Health Sciences, Mailman School of Public Health, Columbia University, New York, NY, USA.(3)Neuromuscular Diseases Research Section, National Institute on Aging, Bethesda, MD, USA.(4)Medical Genetics Laboratory, Department of Pharmacy, Health and Nutritional Sciences, University of Calabria, Rende, Italy.(5)Neuropathology Unit, Department of Brain Sciences, Imperial College London, London, UK.(6)Institute of Psychiatry, Psychology and Neuroscience, King's College London, London, UK.(7)Centre for Age-Related Medicine, Stavanger University Hospital, Stavanger, Norway.(8)Department of Pathology & Laboratory Medicine, School of Medicine, University of California Irvine, Irvine, CA, USA.(9)Department of Pathology and Laboratory Medicine, Indiana University School of Medicine, Indianapolis, IN, USA.(10)Department of Neurology, Oregon Health & Sciences University, Portland, OR, USA.(11)Department of Neurology, Johns Hopkins University Medical Center, Baltimore, MD, USA.(12)Neuroregeneration and Stem Cell Programs, Institute of Cell Engineering, Johns Hopkins University School of Medicine, Baltimore, MD, USA.(13)Department of Pharmacology and Molecular Sciences, Johns Hopkins University School of Medicine, Baltimore, MD, USA.(14)Solomon H. Snyder Department of Neuroscience, Johns Hopkins University School of Medicine, Baltimore, MD, USA.(15)Department of Pathology (Neuropathology), Johns Hopkins University Medical Center, Baltimore, MD, USA.(16)Department of Pathology and Anatomical Sciences, Jacobs School of Medicine and Biomedical Sciences, University at Buffalo, Buffalo, NY, USA.(17)Civin Laboratory for Neuropathology, Banner Sun Health Research Institute, Sun City, AZ, USA.(18)The Sidney Kimmel Comprehensive Cancer Center at Johns Hopkins, Baltimore, MD, USA.(19)Neurodegenerative Diseases Research Unit, National Institute of Neurological Disorders and Stroke, Bethesda, MD, USA. sonja.scholz@nih.gov.(20)Department of Neurology, Johns Hopkins University Medical Center, Baltimore, MD, USA. sonja.scholz@nih.gov.Dementia with Lewy bodies (DLB) is a common form of dementia in the elderly population. We performed genome-wide DNA methylation mapping of cerebellar tissue from pathologically confirmed DLB cases and controls to study the epigenetic profile of this understudied disease. After quality control filtering, 728,197 CpG-sites in 278 cases and 172 controls were available for the analysis. We undertook an epigenome-wide association study, which found a differential methylation signature in DLB cases. Our analysis identified seven differentially methylated probes and three regions associated with DLB. The most significant CpGs were located in ARSB (cg16086807), LINC00173 (cg18800161), and MGRN1 (cg16250093). Functional enrichment evaluations found widespread epigenetic dysregulation in genes associated with neuron-to-neuron synapse, postsynaptic specialization, postsynaptic density, and CTCF-mediated synaptic plasticity. In conclusion, our study highlights the potential importance of epigenetic alterations in the pathogenesis of DLB and provides insights into the modified genes, regions and pathways that may guide therapeutic developments.© 2024. This is a U.S. Government work and not under copyright protection in the US; foreign copyright protection may apply.DOI: 10.1038/s42003-023-05725-xPMCID: PMC10770032PMID: 38182665 [Indexed for MEDLINE]Conflict of interest statement: The authors declare the following competing interests: S.W.S. serves on the Scientific Advisory Council of the Lewy Body Dementia Association and the Multiple System Atrophy Coalition. S.W.S. and B.J.T. receive research support from Cerevel Therapeutics. S.W.S. serves on the editorial board of the journal JAMA Neurology. B.J.T. is associate editor for the journal Brain. B.J.T. holds patents on the clinical testing and therapeutic implication of the C9orf72 repeat expansion. All other authors have no conflicts of interest to declare that are relevant to the content of this article.\"],\n",
       " ['38069335',\n",
       "  '1. Int J Mol Sci. 2023 Nov 30;24(23):17011. doi: 10.3390/ijms242317011.Expression of Tumor Suppressor FHIT Is Regulated by the LINC00173-SNAIL Axis in Human Lung Adenocarcinoma.Suzuki T(1)(2), Sakai S(1), Ota K(1), Yoshida M(1), Uchida C(3), Niida H(1), Suda T(2), Kitagawa M(1), Ohhata T(1).Author information:(1)Department of Molecular Biology, Hamamatsu University School of Medicine, Hamamatsu 431-3192, Japan.(2)Second Division, Department of Internal Medicine, Hamamatsu University School of Medicine, Hamamatsu 431-3192, Japan.(3)Advanced Research Facilities & Services, Preeminent Medical Photonics Education & Research Center, Hamamatsu University School of Medicine, Hamamatsu 431-3192, Japan.Long non-coding RNAs (lncRNAs) play a critical role in a variety of human diseases such as cancer. Here, to elucidate a novel function of a lncRNA called LINC00173, we investigated its binding partner, target gene, and its regulatory mechanism in lung adenocarcinoma, including the A549 cell line and patients. In the A549 cell line, RNA immunoprecipitation (RIP) assays revealed that LINC00173 efficiently binds to SNAIL. RNA-seq and RT-qPCR analyses revealed that the expression of FHIT was decreased upon LINC00173 depletion, indicating that FHIT is a target gene of LINC00173. Overexpression of SNAIL suppressed and depletion of SNAIL increased the expression of FHIT, indicating that SNAIL negatively regulates FHIT. The downregulation of FHIT expression upon LINC00173 depletion was restored by additional SNAIL depletion, revealing a LINC00173-SNAIL-FHIT axis for FHIT regulation. Data from 501 patients with lung adenocarcinoma also support the existence of a LINC00173-SNAIL-FHIT axis, as FHIT expression correlated positively with LINC00173 (p = 1.75 × 10-6) and negatively with SNAIL (p = 7.00 × 10-5). Taken together, we propose that LINC00173 positively regulates FHIT gene expression by binding to SNAIL and inhibiting its function in human lung adenocarcinoma. Thus, this study sheds light on the LINC00173-SNAIL-FHIT axis, which may be a key mechanism for carcinogenesis and progression in human lung adenocarcinoma.DOI: 10.3390/ijms242317011PMCID: PMC10707390PMID: 38069335 [Indexed for MEDLINE]Conflict of interest statement: The authors declare no conflict of interest.'],\n",
       " ['37365917',\n",
       "  \"1. Hum Exp Toxicol. 2023 Jan-Dec;42:9603271231172921. doi: 10.1177/09603271231172921.Long noncoding RNA long intergenic non-protein-coding RNA 173 contributes to nasopharyngeal carcinoma progression by regulating microRNA-765/Gremlin 1 pathway.Wang D(1), Jiang H(1).Author information:(1)Otorhinolaryngologic Department, The Fifth Hospital of Wuhan, Wuhan, China.BACKGROUND: Long intergenic non-protein-coding RNA 173 (LINC00173) executes vital functions in various cancers. Nevertheless, its role and expression in nasopharyngeal carcinoma (NPC) have yet to be investigated. Here, we investigated its effects on the malignancy characteristics of NPC and elucidated the potential molecular mechanism of LINC00173 in NPC progression.METHODS: Quantitative real-time reverse transcription-PCR (qRT-PCR) and immunoblotting were conducted to estimate the LINC00173, microRNA-765 (miR-765), and Gremlin 1 (GREM1) expressions in NPC cells and tissues. Cell counting kit-8 (CCK8), colony formation, and wound healing experiments were done to evaluate the proliferation, growth, and migration of NPC cells, respectively. The tumorous growth of NPC cells in vivo was assessed through the xenograft tumor experiment. Furthermore, the interactions among miR-765, LINC00173, and GREM1 were investigated through bioinformatics analyses, luciferase reporter and RNA immunoprecipitation chip assays.RESULTS: An upregulated LINC00173 expression was found in NPC cell lines and tissues. The functional experiments uncovered that its downregulation repressed NPC cell proliferation, growth, and migration. In addition, LINC00173 knockdown hampered the NPC cells' tumorous growth in vivo. These effects could partially be reversed by downregulating miR-765. GREM1 is a downstream target of miR-765. GREM1 knockdown could repress the proliferation, growth, and migration of NPC cells. Nonetheless, these anti-tumor effects could be abolished by miR-765 downregulation. Mechanistically, LINC00173 increased the expression of GREM1 by binding with miR-765.CONCLUSIONS: LINC00173 functions as an oncogenic factor by binding with miR-765 to promote the progression of NPC via GREM1 upregulation. This study provides a novel insight into the molecular mechanisms involved in NPC progression.DOI: 10.1177/09603271231172921PMID: 37365917 [Indexed for MEDLINE]\"],\n",
       " ['37258811',\n",
       "  \"1. Cancer Gene Ther. 2023 Sep;30(9):1249-1259. doi: 10.1038/s41417-023-00634-x. Epub 2023 May 31.Long noncoding RNA LINC00173 induces radioresistance in nasopharyngeal carcinoma via inhibiting CHK2/P53 pathway.Miao J(#)(1), Chen B(#)(2), Xiao Y(#)(2), Huang R(3), Xiao X(1), Lu S(1), Zhang L(3), Wang X(1), Ouyang Y(2), Chen X(2), Chen Q(1), Xiang Y(1), Guo X(1), Deng X(3), Wang L(4), Mai H(5), Zhao C(6).Author information:(1)Department of Nasopharyngeal Carcinoma, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China.(2)Department of Experimental Research, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China.(3)Department of Radiation Oncology, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China.(4)Department of Nasopharyngeal Carcinoma, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China. wangl1@sysucc.org.cn.(5)Department of Nasopharyngeal Carcinoma, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China. maihq@sysucc.org.cn.(6)Department of Nasopharyngeal Carcinoma, Sun Yat-sen University Cancer Center, State Key Laboratory of Oncology in South China, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Guangzhou, 510060, China. gzzhaochong@hotmail.com.(#)Contributed equallyRadiotherapy is the backbone of nasopharyngeal carcinoma (NPC), nearly 11-17% NPC patients suffered local relapse and 18-37% suffered distant metastasis mainly due to radioresistance. Therefore, the key of improving patients' survivals is to investigate the mechanism of radioresistance. In this study, we revealed that the expression level of long intergenic nonprotein coding RNA 173 (LINC00173) was significantly increased in the radioresistant NPC patients' tumour tissues compared with the radiosensitive patients by RNA-sequencing, which also predict poor prognosis in NPC. Overexpression of LINC00173 induced radioresistance of NPC cells in vitro and in vivo. Mechanistically, LINC00173 bound with checkpoint kinase 2 (CHK2) in nucleus, and impaired the irradiation-induced CHK2 phosphorylation, then suppressed the activation of P53 signalling pathway, which eventually inhibiting apoptosis and leading to radioresistance in NPC cells. In summary, LINC00173 decreases the occurrence of apoptosis through inhibiting the CHK2/P53 pathway, leads to NPC radioresistance and could be considered as a novel predictor and therapeutic target in NPC.© 2023. The Author(s), under exclusive licence to Springer Nature America, Inc.DOI: 10.1038/s41417-023-00634-xPMID: 37258811 [Indexed for MEDLINE]\"],\n",
       " ['37081063',\n",
       "  '1. Sci Rep. 2023 Apr 20;13(1):6492. doi: 10.1038/s41598-023-33059-5.Identification of immune and Toll-like receptor signaling pathway related feature lncRNAs to construct diagnostic nomograms for acute ischemic stroke.Su ZY(1), Yu ZQ(1), Yao B(2), Zhao DX(3).Author information:(1)Changchun University of Chinese Medicine, No.1035 Boshuo Road, Jing Yue National High-Tech Industrial Development Zone, Changchun, 130117, China.(2)School of Aeronautical Fundamentals, Aviation University of Air Force, Changchun, 130041, China.(3)Changchun University of Chinese Medicine, No.1035 Boshuo Road, Jing Yue National High-Tech Industrial Development Zone, Changchun, 130117, China. dexizhao1008@163.com.We aimed to identify the immune and Toll-like receptor (TLR) signaling pathway related feature lncRNAs to construct the diagnostic nomograms for acute ischemic stroke (AIS). Two AIS-associated expression profiles GSE16561 and GSE22255 were downloaded from NCBI Gene Expression Omnibus, the former was the training set and the latter was the validation set. The differential expression genes (DEGs) and lncRNAs (DElncRNAs) related to TLR signaling pathway were identified between AIS and control groups. The single sample gene set enrichment analysis (ssGSEA) was applied to evaluate the immune infiltration. The immune and TLR signaling pathway related DElncRNAs were determined. Three optimization algorithms were utilized to select the immune and TLR signaling pathway related feature lncRNAs to construct the diagnostic nomograms of AIS. Based on the lncRNA signature, a ceRNA network was constructed. 37 DEGs and 28 DElncRNAs related to TLR signaling pathway were identified in GSE16561. 16 immune cell types exhibited significant differences in distribution between AIS and control groups. 28 immune and TLR signaling pathway related DElncRNAs were determined. 8 immune and TLR signaling pathway related feature lncRNAs were selected. The diagnostic nomograms of AIS performed well in both datasets. A ceRNA network was constructed consisting of 7 immune and TLR signaling pathway related feature lncRNAs as well as 19 AIS related miRNAs and 21 TLR signaling pathway related genes. LINC00173, LINC01089, LINC02210, MIR600HG, SNHG14, TP73-AS1, LINC00680 and CASC2 may be the potential biomarkers of AIS diagnosis, and TLR signaling pathway may be a promising immune related therapeutic target for AIS.© 2023. The Author(s).DOI: 10.1038/s41598-023-33059-5PMCID: PMC10119310PMID: 37081063 [Indexed for MEDLINE]Conflict of interest statement: The authors declare no competing interests.'],\n",
       " ['36894691',\n",
       "  '1. Mol Cell Biochem. 2023 Nov;478(11):2553-2565. doi: 10.1007/s11010-023-04684-5. Epub 2023 Mar 9.LncRNA Linc00173 may be a potential prognostic biomarker in human solid tumors: a meta-analysis and bioinformatics analysis.Zhou C(#)(1), Huang Y(#)(1), Tian Y(2), Zhang B(2), Yang X(3)(4)(5)(6).Author information:(1)The First Clinical Medical College of Gansu University of Traditional Chinese Medicine, Lanzhou, 730000, Gansu, China.(2)School of Clinical Medicine, Ningxia Medical University, Yinchuan, 750000, China.(3)Department II of General Surgery, Gansu Provincial Hospital, Lanzhou, 73000, China. yangxjmd@aliyun.com.(4)The First Clinical Medical College of Lanzhou University, Lanzhou University, Lanzhou, 730000, China. yangxjmd@aliyun.com.(5)Gansu Key Laboratory of Molecular Diagnostics and Precision Medicine for Surgical Oncology, Gansu Provincial Hospital, Lanzhou, 730000, China. yangxjmd@aliyun.com.(6)Gansu Research Center of Prevention and Control Project for Digestive Oncology, Gansu Provincial Hospital, Lanzhou, 730000, China. yangxjmd@aliyun.com.(#)Contributed equallyDespite advances in diagnostic techniques and treatments, cancer remains one of the leading causes of death worldwide.Therefore, finding new biomarkers and therapeutic targets is crucial for improving the diagnosis and treatment of human cancer.LncRNA Linc00173 is a newly identified tumor marker, and in this study, we aimed to explore the relationship between Linc00173 and clinicopathological features and patient prognosis. By using The Cochrane Library, EMbase, Web of Science, PubMed, OVID, we conducted a complete and thorough literature search from its inception to November 10, 2022.Meta-analysis was performed using Stata SE16.0 software. Nine studies involving 1102 patients were included.Meta-analysis showed that the overexpression of Linc00173 was significantly associated with poorer OS (HR\\u2009=\\u20091.76,95%CI:1.36-2.26, P\\u2009<\\u20090.001) and shorter DFS (HR\\u2009=\\u20091.89, 95%CI:1.49-2.40,P\\u2009<\\u20090.001),and was significantly associated with gender (male) (OR\\u2009=\\u20091.31,95% CI:1.01-1.69, P\\u2009=\\u20090.042), tumor size (large) (OR\\u2009=\\u20091.34,95% CI:1.01-1.78, P\\u2009=\\u20090.045), and lymph node metastasis (positive) (OR\\u2009=\\u20091.72,95% CI:1.03-2.88, P\\u2009=\\u20090.038). Overexpression of Linc00173 is associated with poor prognosis in cancer patients and is a potential prognostic biomarker and therapeutic target.© 2023. The Author(s), under exclusive licence to Springer Science+Business Media, LLC, part of Springer Nature.DOI: 10.1007/s11010-023-04684-5PMID: 36894691'],\n",
       " ['36856154',\n",
       "  \"1. J Oral Pathol Med. 2023 Jul;52(6):493-503. doi: 10.1111/jop.13425. Epub 2023Mar  14.LINC00173 blocks GATA6-mediated transcription of COL5A1 to affect malignant development of oral squamous cell carcinoma.Dong C(1)(2), Zhao Y(3), Yang S(2), Jiao X(1).Author information:(1)School of Stomatology, Harbin Medical University, Harbin, Heilongjiang, China.(2)Department of Maxillofacial Surgery, Heilongjiang Provincial Hospital, Harbin, Heilongjiang, People's Republic of China.(3)Department of Rehabilitation, The Second Affiliated Hospital of Heilongjiang University of Traditional Chinese Medicine, Harbin, Heilongjiang, China.BACKGROUND: Aberrant expression of collagen type V alpha 1 chain (COL5A1) has been linked to several forms of human cancers. In this work, we focused on the interaction of the LINC00173/GATA binding protein 6 (GATA6)/COL5A1 axis in the malignant property of oral squamous cell carcinoma (OSCC) cells.METHODS: We analyzed six publicly accessible datasets GSE160042, GSE74530, GSE138206, GSE23558, GSE31853 and GSE146483 to identify aberrantly expressed genes in OSCC. The expression of COL5A1 in OSCC tissues and cell lines was examined by reverse transcription-quantitative polymerase chain reaction and/or immunohistochemistry. The regulatory mechanism responsible for COL5A1 transcription was predicted via bioinformatics systems, and the interactions of LINC00173, GATA6, and COL5A1 were identified by immunoprecipitation and luciferase assays. Overexpression or downregulation of COL5A1, GATA6, and LINC00173 were induced in OSCC cell lines to determine their roles in the malignant phenotype of the OSCC cells in vitro and in vivo.RESULTS: COL5A1 showed elevated expression in OSCC tissues and cells. The COLA51 knockdown suppressed proliferation, migration and invasiveness, apoptosis resistance, and pro-angiogenic ability of OSCC cells, and it suppressed the growth and dissemination of xenograft tumors in vivo. GATA6 bound to COL5A1 promoter to activate its transcription, whereas LINC00173 bound to GATA6 to block this transcriptional activation. Overexpression of GATA6 or COL5A1 promoted the malignant phenotype of the OSCC cells, which were blocked upon LINC00173 upregulation.CONCLUSION: This work demonstrates that LINC00173 blocks GATA6-mediated transcription of COL5A1 to affect malignant development of OSCC.© 2023 John Wiley & Sons A/S. Published by John Wiley & Sons Ltd.DOI: 10.1111/jop.13425PMID: 36856154 [Indexed for MEDLINE]\"],\n",
       " ['36774758',\n",
       "  '1. Pathol Res Pract. 2023 Mar;243:154351. doi: 10.1016/j.prp.2023.154351. Epub2023  Feb 3.A review on the role of LINC00173 in human cancers.Ghafouri-Fard S(1), Safarzadeh A(2), Hussen BM(3), Rasul MF(4), Taheri M(5), Akbari Dilmaghani N(6).Author information:(1)Department of Medical Genetics, School of Medicine, Shahid Beheshti University of Medical Sciences, Tehran, Iran.(2)Phytochemistry Research Center, Shahid Beheshti University of Medical Sciences, Tehran, Iran.(3)Department of Pharmacognosy, College of Pharmacy, Hawler Medical University, Kurdistan Region, Erbil, Iraq.(4)Department of Medical Analysis, Faculty of Applied Science, Tishk International University, Kurdistan Region, Erbil, Iraq.(5)Institute of Human Genetics, Jena University Hospital, Jena, Germany; Urology and Nephrology Research Center, Shahid Beheshti University of Medical Sciences, Tehran, Iran. Electronic address: mohammad.taheri@uni-jena.de.(6)Skull Base Research Center, Loghman Hakim Hospital, Shahid Beheshti University of Medical Sciences, Tehran, Iran. Electronic address: nadakbari@sbmu.ac.ir.Long intergenic non-protein coding RNA 173 (LINC00173) is a long non-coding RNA with especial function in the tumorigenic process. Studies in different types of cancers support an oncogenic role for LINC00173 except for studies in B-cell precursor acute lymphoblastic leukemia, cervical cancer, pancreatic cancer and gastric cancer. In breast and lung cancers, both oncogenic and tumor suppressor roles have been reported for LINC00173. LINC00173 can serve as a molecular sponge for miRNAs. miR-218/Etk, miR-511-5p/VEGFA, miR-182-5p/AGER, miR-765/NUTF2, miR-765/PLP2, miR-182-5p/FBXW7, miR-338-3p/Rab25, miR‑641/RAB14 and miR-1275/BCL2 are examples of the miRNA/mRNA axes being regulated by LINC00173 in the context of cancer. The current review provides a summary of different studies on the role of LINC00173 in these cancers.Copyright © 2023 Elsevier GmbH. All rights reserved.DOI: 10.1016/j.prp.2023.154351PMID: 36774758 [Indexed for MEDLINE]Conflict of interest statement: Declaration of Competing Interest The authors declare they have no conflict of interest.'],\n",
       " ['36627670',\n",
       "  \"1. J Exp Clin Cancer Res. 2023 Jan 10;42(1):12. doi: 10.1186/s13046-022-02560-6.Downregulation of Linc00173 increases BCL2 mRNA stability via the miR-1275/PROCA1/ZFP36L2 axis and induces acquired cisplatin resistance of lung adenocarcinoma.Tao X(#)(1), Li Y(#)(1), Fan S(#)(2), Wu L(#)(1), Xin J(1), Su Y(1), Xian X(1), Huang Y(1), Huang R(1), Fang W(3)(4), Liu Z(5)(6).Author information:(1)Guangzhou Municipal and Guangdong Provincial Key Laboratory of Protein Modification and Degradation, State Key Laboratory of Respiratory Disease, Guangzhou Medical University, Guangzhou, 511436, China.(2)The Second Xiangya Hospital of Central South University, Changsha, 410008, China.(3)Cancer Research Institute, School of Basic Medical Sciences, Southern Medical University, Guangzhou, 510515, China. fangweiyi1975@163.com.(4)Cancer Center, Integrated Hospital of Traditional Chinese Medicine, Southern Medical University, Guangzhou, 510315, China. fangweiyi1975@163.com.(5)Cancer Center, Integrated Hospital of Traditional Chinese Medicine, Southern Medical University, Guangzhou, 510315, China. narcissus_jane@163.com.(6)Department of Pathology, Guangzhou Municipal and Guangdong Provincial Key Laboratory of Protein Modification and Degradation, State Key Laboratory of Respiratory Disease, School of Basic Medical Sciences, Guangzhou Medical University, Guangzhou, 511436, China. narcissus_jane@163.com.(#)Contributed equallyBACKGROUND: LINC00173 had been reported as a cisplatin (cis-diamminedichloroplatinum, DDP) chemotherapy-resistant inducer in small-cell lung cancer (SCLC) and lung squamous cell carcinoma (LUSC). This study aimed to display reverse data for LINC00173 as a DDP chemosensitivity-inducing factor in lung adenocarcinoma (LUAD).METHODS: LINC00173 was screened from the Gene Expression Omnibus database (GSE43493). The expression level of LINC00173 in LUAD tissues and cell lines was detected using in situ hybridization and quantitative reverse transcription-polymerase chain reaction. Colony formation, cell viability, half-maximal inhibitory concentration, flow cytometry, and xenograft mouse model were used to evaluate the role of LINC00173 in the chemosensitivity of LUAD to DDP. The mechanism of LINC00173 in DDP resistance by mediating miR-1275/PROCA1/ZFP36L2 axis to impair BCL2 mRNA stability was applied, and co-immunoprecipitation, chromatin immunoprecipitation, RNA antisense purification, RNA immunoprecipitation, and luciferase reporter assays were performed.RESULTS: LINC00173 downregulation in patients with DDP-resistant LUAD was correlated with poor prognosis. Further, LINC00173 expression was significantly reduced in DDP-resistant LUAD cells and DDP-treated human LUAD tissues. Suppressed LINC00173 expression in LUAD cells enhanced DDP chemoresistance in vivo and in vitro, while restored LINC00173 expression in DDP-resistant LUAD cells markedly regained chemosensitivity to DDP. Mechanistically, DDP-resistant LUAD cells activated PI3K/AKT signal and further elevated the c-Myc expression. The c-Myc, as an oncogenic transcriptional factor, bound to the promoter of LINC00173 and suppressed its expression. The reduced LINC00173 expression attenuated the adsorption of oncogenic miR-1275, downregulating the expression of miR-1275 target gene PROCA1. PROCA1 played a potential tumor-suppressive role inducing cell apoptosis and DDP chemosensitivity via recruiting ZFP36L2 to bind to the 3' untranslated region of BCL2, reducing the stability of BCL2 mRNA and thus activating the apoptotic signal.CONCLUSIONS: This study demonstrated a novel and critical role of LINC00173. It was transcriptionally repressed by DDP-activated PI3K/AKT/c-Myc signal in LUAD, promoting DDP-acquired chemotherapeutic resistance by regulating miR-1275 to suppress PROCA1/ZFP36L2-induced BCL2 degradation, which led to apoptotic signal reduction. These data were not consistent with the previously described role of LINC00173 in SCLC or LUSC, which suggested that LINC00173 could play fine-tuned DDP resistance roles in different pathological subtypes of lung cancer. This study demonstrated that the diminished expression of LINC00173 might serve as an indicator of DDP-acquired resistance in LUAD.© 2023. The Author(s).DOI: 10.1186/s13046-022-02560-6PMCID: PMC9830831PMID: 36627670 [Indexed for MEDLINE]Conflict of interest statement: The author declares no competing interest exists.\"],\n",
       " ['36606322',\n",
       "  '1. Mol Oncol. 2023 Mar;17(3):518-533. doi: 10.1002/1878-0261.13375. Epub 2023 Jan 23.LINC00173 facilitates tumor progression by stimulating RAB1B-mediated PA2G4 and SDF4 secretion in nasopharyngeal carcinoma.He SW(1), Liang YL(1), Zhang Y(1), Liu X(1), Gong S(1), Ye ML(1), Huang SY(1), Tan XR(1), Zhou SQ(1), Zhao Y(1), Liu N(1), Li YQ(1).Author information:(1)State Key Laboratory of Oncology in South China, Collaborative Innovation Center of Cancer Medicine, Guangdong Key Laboratory of Nasopharyngeal Carcinoma Diagnosis and Therapy, Sun Yat-sen University Cancer Center, Guangzhou, China.An increasing number of studies have found that long non-coding RNA (lncRNA) play important roles in driving the progression of nasopharyngeal carcinoma (NPC). Our microarray screening revealed that expression of the lncRNA long intergenic non-protein coding RNA 173 (LINC00173) was upregulated in NPC. However, its role and mechanism in NPC have not yet been elucidated. In this study, we demonstrate that high LINC00173 expression indicated a poor prognosis in NPC patients. Knockdown of LINC00173 significantly inhibited NPC cell proliferation, migration and invasion in\\xa0vitro. Mechanistically, LINC00173 interacted and colocalized with Ras-related protein Rab-1B (RAB1B) in the cytoplasm, but the modulation of LINC00173 expression did not affect the expression of RAB1B at either the mRNA or protein levels. Instead, relying on the stimulation of RAB1B, LINC00173 could facilitate the extracellular secretion of proliferation-associated 2G4 (PA2G4) and stromal cell-derived factor 4 (SDF4; also known as 45-kDa calcium-binding protein) proteins, and knockdown of these proteins could reverse the NPC aggressive phenotype induced by LINC00173 overexpression. Moreover, in\\xa0vivo LINC00173-knockdown models exhibited a marked slowdown in tumor growth and a significant reduction in lymph node and lung metastases. In summary, LINC00173 serves as a crucial driver for NPC progression, and the LINC00173-RAB1B-PA2G4/SDF4 axis might provide a potential therapeutic target for NPC patients.© 2023 The Authors. Molecular Oncology published by John Wiley & Sons Ltd on behalf of Federation of European Biochemical Societies.DOI: 10.1002/1878-0261.13375PMCID: PMC9980309PMID: 36606322 [Indexed for MEDLINE]Conflict of interest statement: The authors declare no conflict of interest.']]"
      ]
     },
     "execution_count": 360,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pubmed_res[\"LINC00173\"]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
