{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7809e3d9-3566-4094-8966-eb04f614ae90",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import h5py\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from joblib import Parallel, delayed\n",
    "import subprocess\n",
    "import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5919ca1f-586c-4c75-bdd1-bec7d50e758e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "PWD = Path().cwd()\n",
    "#MAZUR = Path(\"/data/mazurovev/all_marks/methylation/lncRNA_Peaks_corrs\").resolve()\n",
    "DATA = Path(\"../data\")\n",
    "METHYL = DATA / \"methylation_peaks\"\n",
    "METHYL_CONCAT = DATA / \"methylation_peaks_concat\"\n",
    "ANNO = Path(\"/export/home/public/agletdinov_shared/annotations\")\n",
    "PREPROCESSING = Path(\"../preprocessing\")\n",
    "METHYL_INDEX = PREPROCESSING / \"methyl_index\"\n",
    "METHYL_BED = PREPROCESSING / \"methyl_bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1c0706f-5318-4cd9-8d64-fadfaac63a15",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "\n",
    "# Функция для чтения GTF-файла и подсчета уникальных генов\n",
    "def count_genes_in_gtf(gtf_file):\n",
    "    # Открываем файл\n",
    "    with gzip.open(gtf_file, 'rt') as f:\n",
    "        # Считываем строки, относящиеся к генам\n",
    "        gene_lines = [line for line in f if 'gene' in line and line.split('\\t')[2] == 'gene']\n",
    "        \n",
    "    # Выделяем ID генов из строк\n",
    "    gene_ids = set()\n",
    "    for line in gene_lines:\n",
    "        # Разбиваем строку на колонки\n",
    "        columns = line.split('\\t')\n",
    "        # Последняя колонка содержит атрибуты, ищем gene_id\n",
    "        attributes = columns[8]\n",
    "        for attribute in attributes.split(';'):\n",
    "            if 'gene_id' in attribute:\n",
    "                gene_id = attribute.split('\"')[1]\n",
    "                gene_ids.add(gene_id)\n",
    "                break\n",
    "    \n",
    "    return len(gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "469164c3-c853-48fb-9c8c-183ed9b68b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_correlated_peaks(hm, to_hdf5, to_save):\n",
    "    if to_save.exists():\n",
    "        print(f\"Skipped: {to_hdf5.stem}\")\n",
    "        return None\n",
    "    all_results = []\n",
    "\n",
    "    # Открываем HDF5 файл\n",
    "    with h5py.File(to_hdf5, 'r') as f:\n",
    "        lncRNAs = list(f['lncRNAs_names'][:])\n",
    "        corrs_matrix = f['corrs_matrix'][:]\n",
    "\n",
    "    # Читаем пики\n",
    "    peaks = pd.read_csv(Path(\"/data/mazurovev/all_marks\") / hm / \"merged_peaks_first_in_biosample.bed\", sep=\"\\t\", header=None)\n",
    "\n",
    "    for i, lncRNA in enumerate(lncRNAs):\n",
    "        corrs = corrs_matrix[i, :]\n",
    "        nonzero_indices = np.nonzero(corrs)[0]\n",
    "        nonzero_corrs = corrs[nonzero_indices]\n",
    "        nonzero_peaks = [\"peak_\" + str(j) for j in nonzero_indices]\n",
    "\n",
    "        # Извлекаем соответствующие пики\n",
    "        res = peaks[peaks[3].isin(nonzero_peaks)].copy()\n",
    "        res[\"corr\"] = nonzero_corrs\n",
    "        res[\"lnc_ens\"] = lncRNA.decode('utf-8')  # Декодируем байтовую строку в строку\n",
    "\n",
    "        all_results.append(res)\n",
    "\n",
    "    # Объединяем все результаты в один DataFrame\n",
    "    final_result = pd.concat(all_results, ignore_index=True)\n",
    "    \n",
    "    final_result.to_csv(to_save, sep=\"\\t\", compression=\"gzip\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06e5af76-25ab-4aae-8318-3b2e443c7567",
   "metadata": {},
   "source": [
    "for to_hdf5 in MAZUR.glob(\"*.hdf5\"):\n",
    "    print(to_hdf5.stem)\n",
    "    to_save = METHYL / f\"{to_hdf5.stem}.tsv.gz\"\n",
    "    get_all_correlated_peaks(hm=\"methylation\", to_hdf5=to_hdf5, to_save=to_save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f3ec3a7-0ddf-4746-866c-7581c6dced58",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexing_methyl(chunk, lnc_ens, corr_sign, subdf):\n",
    "    to_index = METHYL_INDEX / chunk\n",
    "    to_index.mkdir(exist_ok=True, parents=True)\n",
    "    to_save = to_index / f\"methyl-{lnc_ens}-{corr_sign}-.tsv.gz\"\n",
    "    if to_save.exists():\n",
    "        print(f\"Skipped: methyl {lnc_ens} {corr_sign}\")\n",
    "        return None\n",
    "    subdf.to_csv(to_save, sep=\"\\t\", compression=\"gzip\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4f84cded-b62c-4de9-8909-b9f98e8ec2ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lncRNA_Peaks_Correlations_corrected_non_zero_0.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for to_df in METHYL.glob(\"*.tsv.gz\"):\n",
    "    print(to_df.stem)\n",
    "    chunk = to_df.stem.split(\"_\")[6]\n",
    "    df = pd.read_csv(to_df, sep=\"\\t\")\n",
    "    df.rename(columns={\"0\":\"chrom\", \"1\":\"start\", \"2\":\"end\", \"3\":\"name\"}, inplace=True)\n",
    "    df[\"corr\"] = df[\"corr\"].apply(lambda x: np.round(x, 3))\n",
    "    df[\"corr_sign\"] = np.where(df[\"corr\"] > 0, \"plus\", \"minus\")\n",
    "    for (lnc_ens, corr_sign), subdf in df.groupby(by=[\"lnc_ens\", \"corr_sign\"]):\n",
    "        indexing_methyl(chunk=chunk, lnc_ens=lnc_ens, corr_sign=corr_sign, subdf=subdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abbdbed6-158e-4280-b68c-35e8e72907a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0f50968b-cc45-4133-977c-baba50e6b156",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "/tmp/ipykernel_56201/652964208.py:9: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
      "IOStream.flush timed out\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 10\u001b[0m\n\u001b[1;32m      7\u001b[0m all_peaks \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sign \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mminus\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplus\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m      9\u001b[0m     all_peaks[sign] \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconcat\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mto_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mto_df\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mMETHYL_INDEX\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mglob\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m*\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43msign\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m*\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreset_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdrop\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m\\\u001b[49m\n\u001b[0;32m---> 10\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdrop_duplicates\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mdropna()\n\u001b[1;32m     12\u001b[0m all_peaks_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([all_peaks[sign] \u001b[38;5;28;01mfor\u001b[39;00m sign \u001b[38;5;129;01min\u001b[39;00m (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mminus\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mplus\u001b[39m\u001b[38;5;124m\"\u001b[39m)])\u001b[38;5;241m.\u001b[39mdrop_duplicates()\n\u001b[1;32m     13\u001b[0m \u001b[38;5;28;01mdel\u001b[39;00m all_peaks\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/core/frame.py:6818\u001b[0m, in \u001b[0;36mDataFrame.drop_duplicates\u001b[0;34m(self, subset, keep, inplace, ignore_index)\u001b[0m\n\u001b[1;32m   6815\u001b[0m inplace \u001b[38;5;241m=\u001b[39m validate_bool_kwarg(inplace, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minplace\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6816\u001b[0m ignore_index \u001b[38;5;241m=\u001b[39m validate_bool_kwarg(ignore_index, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mignore_index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 6818\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m[\u001b[38;5;241m-\u001b[39m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mduplicated\u001b[49m\u001b[43m(\u001b[49m\u001b[43msubset\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep\u001b[49m\u001b[43m)\u001b[49m]\n\u001b[1;32m   6819\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ignore_index:\n\u001b[1;32m   6820\u001b[0m     result\u001b[38;5;241m.\u001b[39mindex \u001b[38;5;241m=\u001b[39m default_index(\u001b[38;5;28mlen\u001b[39m(result))\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/core/frame.py:6958\u001b[0m, in \u001b[0;36mDataFrame.duplicated\u001b[0;34m(self, subset, keep)\u001b[0m\n\u001b[1;32m   6956\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6957\u001b[0m     vals \u001b[38;5;241m=\u001b[39m (col\u001b[38;5;241m.\u001b[39mvalues \u001b[38;5;28;01mfor\u001b[39;00m name, col \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mitems() \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m subset)\n\u001b[0;32m-> 6958\u001b[0m     labels, shape \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmap\u001b[39m(\u001b[38;5;28mlist\u001b[39m, \u001b[38;5;28;43mzip\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mmap\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvals\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   6960\u001b[0m     ids \u001b[38;5;241m=\u001b[39m get_group_index(labels, \u001b[38;5;28mtuple\u001b[39m(shape), sort\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, xnull\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m   6961\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor_sliced(duplicated(ids, keep), index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex)\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/core/frame.py:6926\u001b[0m, in \u001b[0;36mDataFrame.duplicated.<locals>.f\u001b[0;34m(vals)\u001b[0m\n\u001b[1;32m   6925\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mf\u001b[39m(vals) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mtuple\u001b[39m[np\u001b[38;5;241m.\u001b[39mndarray, \u001b[38;5;28mint\u001b[39m]:\n\u001b[0;32m-> 6926\u001b[0m     labels, shape \u001b[38;5;241m=\u001b[39m \u001b[43malgorithms\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfactorize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvals\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msize_hint\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6927\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m labels\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mi8\u001b[39m\u001b[38;5;124m\"\u001b[39m, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m), \u001b[38;5;28mlen\u001b[39m(shape)\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/core/algorithms.py:795\u001b[0m, in \u001b[0;36mfactorize\u001b[0;34m(values, sort, use_na_sentinel, size_hint)\u001b[0m\n\u001b[1;32m    792\u001b[0m             \u001b[38;5;66;03m# Don't modify (potentially user-provided) array\u001b[39;00m\n\u001b[1;32m    793\u001b[0m             values \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mwhere(null_mask, na_value, values)\n\u001b[0;32m--> 795\u001b[0m     codes, uniques \u001b[38;5;241m=\u001b[39m \u001b[43mfactorize_array\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    796\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    797\u001b[0m \u001b[43m        \u001b[49m\u001b[43muse_na_sentinel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_na_sentinel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    798\u001b[0m \u001b[43m        \u001b[49m\u001b[43msize_hint\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msize_hint\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    799\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    801\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m sort \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(uniques) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    802\u001b[0m     uniques, codes \u001b[38;5;241m=\u001b[39m safe_sort(\n\u001b[1;32m    803\u001b[0m         uniques,\n\u001b[1;32m    804\u001b[0m         codes,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    807\u001b[0m         verify\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    808\u001b[0m     )\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/core/algorithms.py:595\u001b[0m, in \u001b[0;36mfactorize_array\u001b[0;34m(values, use_na_sentinel, size_hint, na_value, mask)\u001b[0m\n\u001b[1;32m    592\u001b[0m hash_klass, values \u001b[38;5;241m=\u001b[39m _get_hashtable_algo(values)\n\u001b[1;32m    594\u001b[0m table \u001b[38;5;241m=\u001b[39m hash_klass(size_hint \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(values))\n\u001b[0;32m--> 595\u001b[0m uniques, codes \u001b[38;5;241m=\u001b[39m \u001b[43mtable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfactorize\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    596\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    597\u001b[0m \u001b[43m    \u001b[49m\u001b[43mna_sentinel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    598\u001b[0m \u001b[43m    \u001b[49m\u001b[43mna_value\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    599\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    600\u001b[0m \u001b[43m    \u001b[49m\u001b[43mignore_na\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_na_sentinel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    601\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    603\u001b[0m \u001b[38;5;66;03m# re-cast e.g. i8->dt64/td64, uint8->bool\u001b[39;00m\n\u001b[1;32m    604\u001b[0m uniques \u001b[38;5;241m=\u001b[39m _reconstruct_data(uniques, original\u001b[38;5;241m.\u001b[39mdtype, original)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#Не хватает оперативной памяти\n",
    "dtype = {\n",
    "    \"chrom\": str,\n",
    "    \"start\": np.uint32,\n",
    "    \"end\": np.uint32\n",
    "}\n",
    "all_peaks = {}\n",
    "for sign in \"minus\",\"plus\":\n",
    "    all_peaks[sign] = pd.concat([pd.read_csv(to_df, sep=\"\\t\", dtype=dtype, header=None).loc[1:] for to_df in METHYL_INDEX.glob(f\"*{sign}*\")]).reset_index(drop=True)\\\n",
    "    .drop_duplicates().dropna()\n",
    "    \n",
    "all_peaks_df = pd.concat([all_peaks[sign] for sign in (\"minus\",\"plus\")]).drop_duplicates()\n",
    "del all_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bebef528-7e17-43d3-bc13-b70a45d3bd59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr13</td>\n",
       "      <td>99062576</td>\n",
       "      <td>99062577</td>\n",
       "      <td>peak_13294105</td>\n",
       "      <td>-0.855</td>\n",
       "      <td>ENSG00000280047</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>99062606</td>\n",
       "      <td>99062607</td>\n",
       "      <td>peak_13294107</td>\n",
       "      <td>-0.89</td>\n",
       "      <td>ENSG00000280047</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>99062612</td>\n",
       "      <td>99062613</td>\n",
       "      <td>peak_13294109</td>\n",
       "      <td>-0.856</td>\n",
       "      <td>ENSG00000280047</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>32758365</td>\n",
       "      <td>32758366</td>\n",
       "      <td>peak_25313521</td>\n",
       "      <td>-0.869</td>\n",
       "      <td>ENSG00000280047</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0         1         2              3       4                5      6\n",
       "0  chr13  99062576  99062577  peak_13294105  -0.855  ENSG00000280047  minus\n",
       "1  chr13  99062606  99062607  peak_13294107   -0.89  ENSG00000280047  minus\n",
       "2  chr13  99062612  99062613  peak_13294109  -0.856  ENSG00000280047  minus\n",
       "3   chr2  32758365  32758366  peak_25313521  -0.869  ENSG00000280047  minus"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ENSG00000280047 = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*ENSG00000280047*minus*\")]).reset_index(drop=True)\n",
    "ENSG00000280047"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4ac38a86-5c22-43ed-907c-3056c3cca7ea",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n",
      "/tmp/ipykernel_73202/2367621633.py:1: DtypeWarning: Columns (1,2,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "Cell \u001b[0;32mIn[75], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m all_methil_merged \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([\u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mto_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mloc[\u001b[38;5;241m1\u001b[39m:] \u001b[38;5;28;01mfor\u001b[39;00m to_df \u001b[38;5;129;01min\u001b[39;00m METHYL_INDEX\u001b[38;5;241m.\u001b[39mglob(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m*\u001b[39m\u001b[38;5;124m\"\u001b[39m)])\u001b[38;5;241m.\u001b[39mreset_index(drop\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/io/parsers/readers.py:626\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[0;32m--> 626\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1923\u001b[0m, in \u001b[0;36mTextFileReader.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m   1917\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1918\u001b[0m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[1;32m   1919\u001b[0m     (\n\u001b[1;32m   1920\u001b[0m         index,\n\u001b[1;32m   1921\u001b[0m         columns,\n\u001b[1;32m   1922\u001b[0m         col_dict,\n\u001b[0;32m-> 1923\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[1;32m   1924\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[1;32m   1925\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1926\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pandas/io/parsers/c_parser_wrapper.py:234\u001b[0m, in \u001b[0;36mCParserWrapper.read\u001b[0;34m(self, nrows)\u001b[0m\n\u001b[1;32m    233\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlow_memory:\n\u001b[0;32m--> 234\u001b[0m     chunks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reader\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m     \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n",
      "File \u001b[0;32mparsers.pyx:838\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.read_low_memory\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:905\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._read_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:874\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:891\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mparsers.pyx:2053\u001b[0m, in \u001b[0;36mpandas._libs.parsers.raise_parser_error\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/gzip.py:346\u001b[0m, in \u001b[0;36mGzipFile.closed\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    344\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer\u001b[38;5;241m.\u001b[39mpeek(n)\n\u001b[0;32m--> 346\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m    347\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mclosed\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    348\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfileobj \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/IPython/core/interactiveshell.py:3598\u001b[0m, in \u001b[0;36mInteractiveShell.run_code\u001b[0;34m(self, code_obj, result, async_)\u001b[0m\n\u001b[1;32m   3596\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[1;32m   3597\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m-> 3598\u001b[0m         result\u001b[38;5;241m.\u001b[39merror_in_exec \u001b[38;5;241m=\u001b[39m \u001b[43msys\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexc_info\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3599\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshowtraceback(running_compiled_code\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m   3600\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "all_methil_merged = pd.concat([pd.read_csv(to_df, sep=\"\\t\", header=None).loc[1:] for to_df in METHYL_INDEX.glob(\"*\")]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "46848d5f-4405-4a14-b05d-2a455a7b29af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "to_save = METHYL_CONCAT / \"all_methil_merged.tsv.gz\"\n",
    "all_methil_merged.to_csv(to_save, sep=\"\\t\", compression=\"gzip\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f2dea9cc-8f70-42a6-855b-0c30fc6f025c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tsv_to_bed(to_tsv):\n",
    "    to_save = METHYL_BED / f\"{to_tsv.name.split('.')[0]}.bed\"\n",
    "    if to_save.exists():\n",
    "        print(f\"Skipped: {to_save}\")\n",
    "        return None\n",
    "    to_save.parent.mkdir(exist_ok=True, parents=True)\n",
    "    df = pd.read_csv(to_tsv, sep=\"\\t\")\n",
    "    stream = open(to_save, 'w')\n",
    "    for i in range(len(df)):\n",
    "        stream.write(f\"{df.loc[i, 'chrom']}\\t{df.loc[i, 'start']}\\t{df.loc[i, 'end']}\\n\")\n",
    "    \n",
    "    stream.close()\n",
    "    #cmd = f\"gzip {to_save}\" \n",
    "    #subprocess.run(cmd, shell=True, check=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e92daa10-99dc-46d9-b201-ea9eb38d9805",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "with Parallel(n_jobs=-1) as pool:\n",
    "    chunks = pool(\n",
    "        delayed(tsv_to_bed)(to_tsv)\n",
    "        for to_tsv in METHYL_INDEX.glob(\"*\")\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "023ccde3-1f09-49f3-bea8-73f94785dc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_1 = pybedtools.BedTool(METHYL_BED / 'methyl-ENSG00000099869-minus-.bed')\n",
    "test_1 = test_1.sort()\n",
    "test_2 = pybedtools.BedTool(METHYL_BED / 'methyl-ENSG00000103472-minus-.bed')\n",
    "test_2 = test_2.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1ce68f61-af53-4875-8bda-f5ab3fa71252",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr12\t49123587\t49123588\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(test_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9d8f27b6-b1c4-4d9e-9ac3-2a9384a68033",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t15837335\t15837336\n",
      "chr16\t19885468\t19885469\n",
      "chr19\t36064797\t36064798\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(test_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "06efa77d-471b-472d-9017-24c8e2f60e7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t15837335\t15837336\n",
      "chr12\t49123587\t49123588\n",
      "chr16\t19885468\t19885469\n",
      "chr19\t36064797\t36064798\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(test_1.cat(test_2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a6ea7532-b1f8-4e8b-ab63-5f5d96775b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = [i for i in METHYL_BED.glob('*.bed')]\n",
    "x = pybedtools.BedTool(files[0])\n",
    "bed_concat = x.cat(*files[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "166b46f0-d3d2-4db9-a2d2-d817d1ea4b94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17758261"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bed_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "300c1927-00e2-4fb9-b908-f04f7e4e557c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "BEDToolsError",
     "evalue": "\nCommand was:\n\n\tbedtools sort -i /tmp/pybedtools.ei4hweqq.tmp\n\nError message was:\nError: The requested file (/tmp/pybedtools.ei4hweqq.tmp) could not be opened. Error message: (No such file or directory). Exiting!\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mBEDToolsError\u001b[0m                             Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[34], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m bed_concat \u001b[38;5;241m=\u001b[39m \u001b[43mbed_concat\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msort\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pybedtools/bedtool.py:907\u001b[0m, in \u001b[0;36mBedTool._log_to_history.<locals>.decorated\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    903\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdecorated\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    904\u001b[0m \n\u001b[1;32m    905\u001b[0m     \u001b[38;5;66;03m# this calls the actual method in the first place; *result* is\u001b[39;00m\n\u001b[1;32m    906\u001b[0m     \u001b[38;5;66;03m# whatever you get back\u001b[39;00m\n\u001b[0;32m--> 907\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    909\u001b[0m     \u001b[38;5;66;03m# add appropriate tags\u001b[39;00m\n\u001b[1;32m    910\u001b[0m     parent_tag \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_tag\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pybedtools/bedtool.py:388\u001b[0m, in \u001b[0;36m_wraps.<locals>.decorator.<locals>.wrapped\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    385\u001b[0m decode_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m result_is_bam\n\u001b[1;32m    387\u001b[0m \u001b[38;5;66;03m# Do the actual call\u001b[39;00m\n\u001b[0;32m--> 388\u001b[0m stream \u001b[38;5;241m=\u001b[39m \u001b[43mcall_bedtools\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    389\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcmds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    390\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtmp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    391\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstdin\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstdin\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    392\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcheck_stderr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcheck_stderr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    393\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdecode_output\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecode_output\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    394\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    396\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m does_not_return_bedtool:\n\u001b[1;32m    397\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m does_not_return_bedtool(stream, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/miniforge3/envs/work/lib/python3.12/site-packages/pybedtools/helpers.py:456\u001b[0m, in \u001b[0;36mcall_bedtools\u001b[0;34m(cmds, tmpfn, stdin, check_stderr, decode_output, encode_input)\u001b[0m\n\u001b[1;32m    454\u001b[0m             sys\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mwrite(stderr)\n\u001b[1;32m    455\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 456\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m BEDToolsError(subprocess\u001b[38;5;241m.\u001b[39mlist2cmdline(cmds), stderr)\n\u001b[1;32m    458\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mOSError\u001b[39;00m, \u001b[38;5;167;01mIOError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m    459\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mtype\u001b[39m(err), os\u001b[38;5;241m.\u001b[39mstrerror(err\u001b[38;5;241m.\u001b[39merrno)))\n",
      "\u001b[0;31mBEDToolsError\u001b[0m: \nCommand was:\n\n\tbedtools sort -i /tmp/pybedtools.ei4hweqq.tmp\n\nError message was:\nError: The requested file (/tmp/pybedtools.ei4hweqq.tmp) could not be opened. Error message: (No such file or directory). Exiting!\n"
     ]
    }
   ],
   "source": [
    "bed_concat = bed_concat.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "522f20f7-93d1-4786-85d5-b9c3689ba99d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(../preprocessing/methyl_bed/methyl_concat/methyl_concat.bed)>"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_concat.moveto(METHYL_BED / \"methyl_concat\" /'methyl_concat.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "30f0637a-0d93-461e-bb74-590cbd29dfcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "peaks_anno_df = pd.read_csv(METHYL_BED / \"methyl_concat/peaks_anno.csv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4d9629c2-47dc-4a07-a58e-e02c3a2dd35e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak</th>\n",
       "      <th>feature</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>feature_strand</th>\n",
       "      <th>insideFeature</th>\n",
       "      <th>distancetoFeature</th>\n",
       "      <th>shortestDistance</th>\n",
       "      <th>fromOverlappingOrNearest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10542</td>\n",
       "      <td>10542</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10571</td>\n",
       "      <td>10571</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10577</td>\n",
       "      <td>10577</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10579</td>\n",
       "      <td>10579</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10610</td>\n",
       "      <td>10610</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqnames  start    end  width strand       peak feature  start_position  \\\n",
       "1     chr1  10542  10542      1      *  X00000001     NaN             NaN   \n",
       "2     chr1  10571  10571      1      *  X00000002     NaN             NaN   \n",
       "3     chr1  10577  10577      1      *  X00000003     NaN             NaN   \n",
       "4     chr1  10579  10579      1      *  X00000004     NaN             NaN   \n",
       "5     chr1  10610  10610      1      *  X00000005     NaN             NaN   \n",
       "\n",
       "   end_position feature_strand insideFeature  distancetoFeature  \\\n",
       "1           NaN            NaN           NaN                NaN   \n",
       "2           NaN            NaN           NaN                NaN   \n",
       "3           NaN            NaN           NaN                NaN   \n",
       "4           NaN            NaN           NaN                NaN   \n",
       "5           NaN            NaN           NaN                NaN   \n",
       "\n",
       "   shortestDistance fromOverlappingOrNearest  \n",
       "1               NaN                      NaN  \n",
       "2               NaN                      NaN  \n",
       "3               NaN                      NaN  \n",
       "4               NaN                      NaN  \n",
       "5               NaN                      NaN  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_anno_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b7329934-a006-4853-81ad-46c997effbe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = peaks_anno_df[\"feature\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3435ce14-25d3-478a-ad65-0c2a55760e8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan, 'ENSG00000225880.4', 'ENSG00000240453.1', 'CATG00000042729.1',\n",
       "       'CATG00000042732.1', 'CATG00000042734.1', 'ENSG00000228327.2',\n",
       "       'ENSG00000237491.4', 'ENSG00000177757.1', 'ENSG00000228794.4'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a7d3305e-df64-4ff6-9527-82d6787ebe57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>peak</th>\n",
       "      <th>feature</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>feature_strand</th>\n",
       "      <th>insideFeature</th>\n",
       "      <th>distancetoFeature</th>\n",
       "      <th>shortestDistance</th>\n",
       "      <th>fromOverlappingOrNearest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>91684</td>\n",
       "      <td>91685</td>\n",
       "      <td>2</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000182</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>671202.0</td>\n",
       "      <td>263.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92392</td>\n",
       "      <td>92392</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000183</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>670494.0</td>\n",
       "      <td>971.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92742</td>\n",
       "      <td>92742</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000184</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>670144.0</td>\n",
       "      <td>1321.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>chr1</td>\n",
       "      <td>92763</td>\n",
       "      <td>92763</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000185</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>670123.0</td>\n",
       "      <td>1342.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>chr1</td>\n",
       "      <td>93393</td>\n",
       "      <td>93393</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00000186</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>669493.0</td>\n",
       "      <td>1972.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3304</th>\n",
       "      <td>chr1</td>\n",
       "      <td>759226</td>\n",
       "      <td>759226</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00001351</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>3660.0</td>\n",
       "      <td>3660.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>chr1</td>\n",
       "      <td>759318</td>\n",
       "      <td>759318</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00001352</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>3568.0</td>\n",
       "      <td>3568.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3310</th>\n",
       "      <td>chr1</td>\n",
       "      <td>759876</td>\n",
       "      <td>759876</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00001353</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>3010.0</td>\n",
       "      <td>3010.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3313</th>\n",
       "      <td>chr1</td>\n",
       "      <td>761800</td>\n",
       "      <td>761801</td>\n",
       "      <td>2</td>\n",
       "      <td>*</td>\n",
       "      <td>X00001354</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>1086.0</td>\n",
       "      <td>1085.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>chr1</td>\n",
       "      <td>761890</td>\n",
       "      <td>761890</td>\n",
       "      <td>1</td>\n",
       "      <td>*</td>\n",
       "      <td>X00001355</td>\n",
       "      <td>ENSG00000225880.4</td>\n",
       "      <td>91421.0</td>\n",
       "      <td>762886.0</td>\n",
       "      <td>-</td>\n",
       "      <td>inside</td>\n",
       "      <td>996.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>Overlapping</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1174 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqnames   start     end  width strand       peak            feature  \\\n",
       "182      chr1   91684   91685      2      *  X00000182  ENSG00000225880.4   \n",
       "184      chr1   92392   92392      1      *  X00000183  ENSG00000225880.4   \n",
       "186      chr1   92742   92742      1      *  X00000184  ENSG00000225880.4   \n",
       "188      chr1   92763   92763      1      *  X00000185  ENSG00000225880.4   \n",
       "190      chr1   93393   93393      1      *  X00000186  ENSG00000225880.4   \n",
       "...       ...     ...     ...    ...    ...        ...                ...   \n",
       "3304     chr1  759226  759226      1      *  X00001351  ENSG00000225880.4   \n",
       "3307     chr1  759318  759318      1      *  X00001352  ENSG00000225880.4   \n",
       "3310     chr1  759876  759876      1      *  X00001353  ENSG00000225880.4   \n",
       "3313     chr1  761800  761801      2      *  X00001354  ENSG00000225880.4   \n",
       "3317     chr1  761890  761890      1      *  X00001355  ENSG00000225880.4   \n",
       "\n",
       "      start_position  end_position feature_strand insideFeature  \\\n",
       "182          91421.0      762886.0              -        inside   \n",
       "184          91421.0      762886.0              -        inside   \n",
       "186          91421.0      762886.0              -        inside   \n",
       "188          91421.0      762886.0              -        inside   \n",
       "190          91421.0      762886.0              -        inside   \n",
       "...              ...           ...            ...           ...   \n",
       "3304         91421.0      762886.0              -        inside   \n",
       "3307         91421.0      762886.0              -        inside   \n",
       "3310         91421.0      762886.0              -        inside   \n",
       "3313         91421.0      762886.0              -        inside   \n",
       "3317         91421.0      762886.0              -        inside   \n",
       "\n",
       "      distancetoFeature  shortestDistance fromOverlappingOrNearest  \n",
       "182            671202.0             263.0              Overlapping  \n",
       "184            670494.0             971.0              Overlapping  \n",
       "186            670144.0            1321.0              Overlapping  \n",
       "188            670123.0            1342.0              Overlapping  \n",
       "190            669493.0            1972.0              Overlapping  \n",
       "...                 ...               ...                      ...  \n",
       "3304             3660.0            3660.0              Overlapping  \n",
       "3307             3568.0            3568.0              Overlapping  \n",
       "3310             3010.0            3010.0              Overlapping  \n",
       "3313             1086.0            1085.0              Overlapping  \n",
       "3317              996.0             996.0              Overlapping  \n",
       "\n",
       "[1174 rows x 14 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_anno_df[peaks_anno_df[\"feature\"] == 'ENSG00000225880.4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3062aecf-2182-4fb4-ae0c-7cf59414db4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "668187"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "759608 - 91421"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "67fa5bcf-46f6-4ac2-9e87-0f2f1116169b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-33"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "762886 - 762919"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84670e6d-306a-49af-a0de-d32838a9984a",
   "metadata": {},
   "outputs": [],
   "source": [
    "762919"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37797cb0-a6ca-472b-8e98-51bde599bb97",
   "metadata": {},
   "outputs": [],
   "source": [
    "91421.0\t762886.0\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4a1f168c-dc8f-421e-93d2-f686dd46961d",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_csv= METHYL_CONCAT / \"all_methil_merged.tsv.gz\"\n",
    "all_methil_merged = pd.read_csv(to_csv, sep=\"\\t\", compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "21c60a5e-280f-4e57-be0a-be87fd880bc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>65540025</td>\n",
       "      <td>65540026</td>\n",
       "      <td>peak_8221299</td>\n",
       "      <td>0.824</td>\n",
       "      <td>ENSG00000274317</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>88534857</td>\n",
       "      <td>88534858</td>\n",
       "      <td>peak_18894879</td>\n",
       "      <td>0.852</td>\n",
       "      <td>ENSG00000274317</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>74861803</td>\n",
       "      <td>74861804</td>\n",
       "      <td>peak_20845738</td>\n",
       "      <td>0.804</td>\n",
       "      <td>ENSG00000274317</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr19</td>\n",
       "      <td>4911993</td>\n",
       "      <td>4911994</td>\n",
       "      <td>peak_22893683</td>\n",
       "      <td>0.831</td>\n",
       "      <td>ENSG00000274317</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr21</td>\n",
       "      <td>35046939</td>\n",
       "      <td>35046940</td>\n",
       "      <td>peak_30600780</td>\n",
       "      <td>0.867</td>\n",
       "      <td>ENSG00000274317</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825392</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.594</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825393</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152353</td>\n",
       "      <td>71152354</td>\n",
       "      <td>peak_45219204</td>\n",
       "      <td>-0.636</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825394</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152851</td>\n",
       "      <td>71152852</td>\n",
       "      <td>peak_45219214</td>\n",
       "      <td>-0.573</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825395</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152920</td>\n",
       "      <td>71152921</td>\n",
       "      <td>peak_45219216</td>\n",
       "      <td>-0.600</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825396</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71153306</td>\n",
       "      <td>71153307</td>\n",
       "      <td>peak_45219227</td>\n",
       "      <td>-0.616</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>277825397 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0         1         2              3      4                5  \\\n",
       "0          chr11  65540025  65540026   peak_8221299  0.824  ENSG00000274317   \n",
       "1          chr16  88534857  88534858  peak_18894879  0.852  ENSG00000274317   \n",
       "2          chr17  74861803  74861804  peak_20845738  0.804  ENSG00000274317   \n",
       "3          chr19   4911993   4911994  peak_22893683  0.831  ENSG00000274317   \n",
       "4          chr21  35046939  35046940  peak_30600780  0.867  ENSG00000274317   \n",
       "...          ...       ...       ...            ...    ...              ...   \n",
       "277825392   chr7  71152278  71152279  peak_45219202 -0.594  ENSG00000273080   \n",
       "277825393   chr7  71152353  71152354  peak_45219204 -0.636  ENSG00000273080   \n",
       "277825394   chr7  71152851  71152852  peak_45219214 -0.573  ENSG00000273080   \n",
       "277825395   chr7  71152920  71152921  peak_45219216 -0.600  ENSG00000273080   \n",
       "277825396   chr7  71153306  71153307  peak_45219227 -0.616  ENSG00000273080   \n",
       "\n",
       "               6  \n",
       "0           plus  \n",
       "1           plus  \n",
       "2           plus  \n",
       "3           plus  \n",
       "4           plus  \n",
       "...          ...  \n",
       "277825392  minus  \n",
       "277825393  minus  \n",
       "277825394  minus  \n",
       "277825395  minus  \n",
       "277825396  minus  \n",
       "\n",
       "[277825397 rows x 7 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6e7e9119-af34-451f-8466-d13454bc30e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(277825397, 7)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged_filt = all_methil_merged.drop_duplicates(subset=[\"0\",\"1\",\"2\",\"5\"])\n",
    "all_methil_merged_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "161e3d6d-76ba-4e76-879a-0f0c136fbf57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25177317, 7)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged_filt = all_methil_merged.drop_duplicates(subset=[\"0\",\"1\",\"2\"])\n",
    "all_methil_merged_filt.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "14d1beea-3205-452b-b4fc-49d94fd51838",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [0, 1, 2, 3, 4, 5, 6]\n",
       "Index: []"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = all_methil_merged[\"5\"].apply(lambda x: \"ENSG00000225880\" in x)\n",
    "all_methil_merged[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "057ecc59-530d-4ee1-961c-1e9c42d7de5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(277825397, 7)"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "c48a92b6-23f8-4948-843f-ff52808d1ed9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0               chr7\n",
       "1           71152278\n",
       "2           71152279\n",
       "3      peak_45219202\n",
       "4             -0.594\n",
       "5    ENSG00000273080\n",
       "6              minus\n",
       "Name: 277825392, dtype: object"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged.loc[277825392]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "2c2cd39d-8b6e-4867-ba89-8df453b2ce57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>249128659</th>\n",
       "      <td>chr4</td>\n",
       "      <td>10541</td>\n",
       "      <td>10542</td>\n",
       "      <td>peak_35232603</td>\n",
       "      <td>-0.543</td>\n",
       "      <td>ENSG00000278989</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0      1      2              3      4                5      6\n",
       "249128659  chr4  10541  10542  peak_35232603 -0.543  ENSG00000278989  minus"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged[all_methil_merged[\"1\"] == 10541]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "c14e02a0-2d8b-4ad6-8d6f-5cf1dee8e6b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3484824</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.634</td>\n",
       "      <td>ENSG00000267506</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6831216</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.642</td>\n",
       "      <td>ENSG00000248441</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8359489</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.657</td>\n",
       "      <td>ENSG00000227888</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12861960</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.742</td>\n",
       "      <td>ENSG00000258498</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15595068</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.619</td>\n",
       "      <td>ENSG00000254614</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23519914</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.616</td>\n",
       "      <td>ENSG00000256576</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26166893</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.686</td>\n",
       "      <td>ENSG00000257913</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35300918</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.611</td>\n",
       "      <td>ENSG00000228649</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40754533</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.564</td>\n",
       "      <td>ENSG00000280234</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48872330</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.667</td>\n",
       "      <td>ENSG00000267272</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51519863</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.620</td>\n",
       "      <td>ENSG00000231170</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53710177</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.658</td>\n",
       "      <td>ENSG00000279594</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57893798</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.554</td>\n",
       "      <td>ENSG00000276724</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61383017</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.752</td>\n",
       "      <td>ENSG00000226194</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79737408</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.581</td>\n",
       "      <td>ENSG00000242125</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84411008</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.728</td>\n",
       "      <td>ENSG00000273987</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90752619</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.694</td>\n",
       "      <td>ENSG00000279020</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97717525</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.635</td>\n",
       "      <td>ENSG00000267784</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102607010</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.625</td>\n",
       "      <td>ENSG00000228956</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104683202</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.711</td>\n",
       "      <td>ENSG00000163597</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110184886</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.635</td>\n",
       "      <td>ENSG00000245105</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127510176</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.551</td>\n",
       "      <td>ENSG00000280042</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132613209</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.624</td>\n",
       "      <td>ENSG00000233008</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136153968</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.629</td>\n",
       "      <td>ENSG00000279099</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141084741</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.628</td>\n",
       "      <td>ENSG00000277954</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145730724</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.765</td>\n",
       "      <td>ENSG00000263069</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158956212</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.511</td>\n",
       "      <td>ENSG00000230337</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161607491</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.643</td>\n",
       "      <td>ENSG00000235023</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175892214</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.803</td>\n",
       "      <td>ENSG00000273186</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188241891</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.631</td>\n",
       "      <td>ENSG00000231160</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196619468</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.589</td>\n",
       "      <td>ENSG00000260816</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201247911</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.741</td>\n",
       "      <td>ENSG00000226091</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204226026</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.675</td>\n",
       "      <td>ENSG00000251034</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212677904</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.704</td>\n",
       "      <td>ENSG00000280169</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219711805</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.696</td>\n",
       "      <td>ENSG00000259495</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226603678</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.543</td>\n",
       "      <td>ENSG00000280237</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235924567</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.602</td>\n",
       "      <td>ENSG00000245573</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241010812</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.628</td>\n",
       "      <td>ENSG00000255121</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245827179</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.629</td>\n",
       "      <td>ENSG00000261126</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248486751</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.565</td>\n",
       "      <td>ENSG00000279254</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256846757</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>0.603</td>\n",
       "      <td>ENSG00000238039</td>\n",
       "      <td>plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267945349</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.610</td>\n",
       "      <td>ENSG00000260778</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277825392</th>\n",
       "      <td>chr7</td>\n",
       "      <td>71152278</td>\n",
       "      <td>71152279</td>\n",
       "      <td>peak_45219202</td>\n",
       "      <td>-0.594</td>\n",
       "      <td>ENSG00000273080</td>\n",
       "      <td>minus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0         1         2              3      4                5  \\\n",
       "3484824    chr7  71152278  71152279  peak_45219202  0.634  ENSG00000267506   \n",
       "6831216    chr7  71152278  71152279  peak_45219202  0.642  ENSG00000248441   \n",
       "8359489    chr7  71152278  71152279  peak_45219202  0.657  ENSG00000227888   \n",
       "12861960   chr7  71152278  71152279  peak_45219202  0.742  ENSG00000258498   \n",
       "15595068   chr7  71152278  71152279  peak_45219202 -0.619  ENSG00000254614   \n",
       "23519914   chr7  71152278  71152279  peak_45219202  0.616  ENSG00000256576   \n",
       "26166893   chr7  71152278  71152279  peak_45219202 -0.686  ENSG00000257913   \n",
       "35300918   chr7  71152278  71152279  peak_45219202 -0.611  ENSG00000228649   \n",
       "40754533   chr7  71152278  71152279  peak_45219202  0.564  ENSG00000280234   \n",
       "48872330   chr7  71152278  71152279  peak_45219202  0.667  ENSG00000267272   \n",
       "51519863   chr7  71152278  71152279  peak_45219202  0.620  ENSG00000231170   \n",
       "53710177   chr7  71152278  71152279  peak_45219202  0.658  ENSG00000279594   \n",
       "57893798   chr7  71152278  71152279  peak_45219202  0.554  ENSG00000276724   \n",
       "61383017   chr7  71152278  71152279  peak_45219202  0.752  ENSG00000226194   \n",
       "79737408   chr7  71152278  71152279  peak_45219202 -0.581  ENSG00000242125   \n",
       "84411008   chr7  71152278  71152279  peak_45219202  0.728  ENSG00000273987   \n",
       "90752619   chr7  71152278  71152279  peak_45219202  0.694  ENSG00000279020   \n",
       "97717525   chr7  71152278  71152279  peak_45219202  0.635  ENSG00000267784   \n",
       "102607010  chr7  71152278  71152279  peak_45219202  0.625  ENSG00000228956   \n",
       "104683202  chr7  71152278  71152279  peak_45219202 -0.711  ENSG00000163597   \n",
       "110184886  chr7  71152278  71152279  peak_45219202  0.635  ENSG00000245105   \n",
       "127510176  chr7  71152278  71152279  peak_45219202  0.551  ENSG00000280042   \n",
       "132613209  chr7  71152278  71152279  peak_45219202  0.624  ENSG00000233008   \n",
       "136153968  chr7  71152278  71152279  peak_45219202  0.629  ENSG00000279099   \n",
       "141084741  chr7  71152278  71152279  peak_45219202  0.628  ENSG00000277954   \n",
       "145730724  chr7  71152278  71152279  peak_45219202  0.765  ENSG00000263069   \n",
       "158956212  chr7  71152278  71152279  peak_45219202 -0.511  ENSG00000230337   \n",
       "161607491  chr7  71152278  71152279  peak_45219202  0.643  ENSG00000235023   \n",
       "175892214  chr7  71152278  71152279  peak_45219202 -0.803  ENSG00000273186   \n",
       "188241891  chr7  71152278  71152279  peak_45219202  0.631  ENSG00000231160   \n",
       "196619468  chr7  71152278  71152279  peak_45219202  0.589  ENSG00000260816   \n",
       "201247911  chr7  71152278  71152279  peak_45219202  0.741  ENSG00000226091   \n",
       "204226026  chr7  71152278  71152279  peak_45219202  0.675  ENSG00000251034   \n",
       "212677904  chr7  71152278  71152279  peak_45219202  0.704  ENSG00000280169   \n",
       "219711805  chr7  71152278  71152279  peak_45219202  0.696  ENSG00000259495   \n",
       "226603678  chr7  71152278  71152279  peak_45219202  0.543  ENSG00000280237   \n",
       "235924567  chr7  71152278  71152279  peak_45219202  0.602  ENSG00000245573   \n",
       "241010812  chr7  71152278  71152279  peak_45219202 -0.628  ENSG00000255121   \n",
       "245827179  chr7  71152278  71152279  peak_45219202  0.629  ENSG00000261126   \n",
       "248486751  chr7  71152278  71152279  peak_45219202  0.565  ENSG00000279254   \n",
       "256846757  chr7  71152278  71152279  peak_45219202  0.603  ENSG00000238039   \n",
       "267945349  chr7  71152278  71152279  peak_45219202 -0.610  ENSG00000260778   \n",
       "277825392  chr7  71152278  71152279  peak_45219202 -0.594  ENSG00000273080   \n",
       "\n",
       "               6  \n",
       "3484824     plus  \n",
       "6831216     plus  \n",
       "8359489     plus  \n",
       "12861960    plus  \n",
       "15595068   minus  \n",
       "23519914    plus  \n",
       "26166893   minus  \n",
       "35300918   minus  \n",
       "40754533    plus  \n",
       "48872330    plus  \n",
       "51519863    plus  \n",
       "53710177    plus  \n",
       "57893798    plus  \n",
       "61383017    plus  \n",
       "79737408   minus  \n",
       "84411008    plus  \n",
       "90752619    plus  \n",
       "97717525    plus  \n",
       "102607010   plus  \n",
       "104683202  minus  \n",
       "110184886   plus  \n",
       "127510176   plus  \n",
       "132613209   plus  \n",
       "136153968   plus  \n",
       "141084741   plus  \n",
       "145730724   plus  \n",
       "158956212  minus  \n",
       "161607491   plus  \n",
       "175892214  minus  \n",
       "188241891   plus  \n",
       "196619468   plus  \n",
       "201247911   plus  \n",
       "204226026   plus  \n",
       "212677904   plus  \n",
       "219711805   plus  \n",
       "226603678   plus  \n",
       "235924567   plus  \n",
       "241010812  minus  \n",
       "245827179   plus  \n",
       "248486751   plus  \n",
       "256846757   plus  \n",
       "267945349  minus  \n",
       "277825392  minus  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_methil_merged[all_methil_merged[\"3\"] == \"peak_45219202\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c939f06f-a654-4527-bd17-d7f6f74478f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "eb9ec3e8-a601-471a-af7b-23a2f8b82f6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(25200691, 14)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_anno_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3beb5308-f01c-426c-917a-d80e117ed073",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_concat = pybedtools.BedTool(METHYL_BED / \"methyl_concat\" /'methyl_concat.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "6aa6151b-12b0-4448-a115-6c69721befaf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17758261"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bed_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "617e6723-64ac-4e7c-bc72-725ee14b68d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "methyl_peaks = pd.merge(left=all_methil_merged, left_on=\"target.gapmerID\",\n",
    "         right=fibro_aso[[\"aso_id\", \"target.geneID\"]], right_on=\"aso_id\")\n",
    "fibro_expr_filt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91421f70-1e99-4ba0-8888-fcfb40519193",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4a04af11-8f6c-4f41-be40-a1fda99bfb4d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lncRNA_Peaks_Correlations_corrected_non_zero_0.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_10.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_11.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_12.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_13.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_14.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_15.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_16.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_17.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_18.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_1.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_2.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_3.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_4.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_5.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_6.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_7.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_8.hdf5\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_9.hdf5\n"
     ]
    }
   ],
   "source": [
    "!ls /data/mazurovev/all_marks/methylation/lncRNA_Peaks_corrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7ef7a050-0946-4945-b93f-b9472ec2a73e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lncRNA_Peaks_Correlations_corrected_non_zero_0.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_10.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_11.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_12.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_13.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_14.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_15.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_16.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_17.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_18.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_1.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_2.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_3.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_4.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_5.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_6.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_7.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_8.tsv.gz\n",
      "lncRNA_Peaks_Correlations_corrected_non_zero_9.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "!ls {METHYL}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa21dc64-0563-4404-8e46-1e7ecf225703",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91dc32a6-9500-4859-a5ec-30079da2bfc6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a9d9d50-8cae-437c-850c-426526b95bf1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b985eb2d-3312-41de-a347-1d8a1d6d2c42",
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_df = pd.concat([pd.read_csv(to_df, sep=\"\\t\") for to_df in METHYL.glob(\"*.tsv.gz\")])\n",
    "corr_df.sort_values(by=[\"lnc_ens\", \"0\", \"1\"], inplace=True)\n",
    "corr_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ea22f48-53ad-4907-9c72-f43e2c7585ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9eda311-3beb-4542-a6ff-8e9df87c96ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "b18f2155-90a5-499b-bda7-a7908a97ad6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5181581, 9)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Указываем путь к файлу\n",
    "gtf_file = ANNO / \"FANTOM_CAT.lv2_permissive.gtf.gz\"\n",
    "anno = pd.read_csv(gtf_file, sep=\"\\t\", header=None)\n",
    "anno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "0ad6ed41-967e-4ff8-9452-92e98d506fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = anno[8].apply(lambda x: \"ENSG00000225880\" in x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "6fd9ec86-a6fc-4d1c-8ffc-5d7a50513ac7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4778223</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>gene</td>\n",
       "      <td>91421</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; geneSuperClass \"a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778224</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>761586</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778225</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>761586</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778226</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>759426</td>\n",
       "      <td>762598</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778227</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>759426</td>\n",
       "      <td>762598</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778228</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>91421</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778229</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>91421</td>\n",
       "      <td>91629</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778230</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>92091</td>\n",
       "      <td>92240</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778231</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>259017</td>\n",
       "      <td>259121</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778232</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>655412</td>\n",
       "      <td>655580</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778233</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>741179</td>\n",
       "      <td>741271</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778234</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>762864</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"FT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778235</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>transcript</td>\n",
       "      <td>761589</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4778236</th>\n",
       "      <td>chr1</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>exon</td>\n",
       "      <td>761589</td>\n",
       "      <td>762886</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225880.4\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0       1           2       3       4  5  6  7  \\\n",
       "4778223  chr1  FANTOM        gene   91421  762886  .  -  .   \n",
       "4778224  chr1  FANTOM  transcript  761586  762886  .  -  .   \n",
       "4778225  chr1  FANTOM        exon  761586  762886  .  -  .   \n",
       "4778226  chr1  FANTOM  transcript  759426  762598  .  -  .   \n",
       "4778227  chr1  FANTOM        exon  759426  762598  .  -  .   \n",
       "4778228  chr1  FANTOM  transcript   91421  762886  .  -  .   \n",
       "4778229  chr1  FANTOM        exon   91421   91629  .  -  .   \n",
       "4778230  chr1  FANTOM        exon   92091   92240  .  -  .   \n",
       "4778231  chr1  FANTOM        exon  259017  259121  .  -  .   \n",
       "4778232  chr1  FANTOM        exon  655412  655580  .  -  .   \n",
       "4778233  chr1  FANTOM        exon  741179  741271  .  -  .   \n",
       "4778234  chr1  FANTOM        exon  762864  762886  .  -  .   \n",
       "4778235  chr1  FANTOM  transcript  761589  762886  .  -  .   \n",
       "4778236  chr1  FANTOM        exon  761589  762886  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "4778223  gene_id \"ENSG00000225880.4\"; geneSuperClass \"a...  \n",
       "4778224  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  \n",
       "4778225  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  \n",
       "4778226  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  \n",
       "4778227  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  \n",
       "4778228  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778229  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778230  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778231  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778232  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778233  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778234  gene_id \"ENSG00000225880.4\"; transcript_id \"FT...  \n",
       "4778235  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  \n",
       "4778236  gene_id \"ENSG00000225880.4\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d953a01-f765-468d-bd9c-c4c02216857e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e530750d-d052-4766-a5cc-f453a45bae48",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
